---
title: "NRCS SHAI Dataset"
author: "Dan Manter"
date: "`r format(Sys.time(), '%m/%d/%y')`"
output: bookdown::html_document2
---

# This file is for inital processing of the SHAI dataset. 
# Calculates PLFA indicator groups
# Calculates Soil Health Management Index (SHMI)
# Downloads DayMet weather data for each location
# Adds taxa and PICRUSt2 data to files
# Combines all data into Excel and Phyloseq objects for downstream analysis
#
# Meta data, CASH, and PLFA are already in source file:
# '../data/raw_data/myPhyloDB/SHAI.Meta.7Feb2022.RelAbund.xlsm'
# 
# Final Output Files
# Excel
# '../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.KO.RelAbund.xlsx' 
#                 --- PLFA mol%
#                 --- DNA rel abund (only GIBBs are normalized by 16S & func gene copies))
# '../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.KO.Total.xlsx'
#                 --- PLFA pmol g-1 soil
#                 --- DNA copies g-1 soil (only GIBBs are normalized by 16S & func gene copies)
# 
# Phyloseq Files (DNA)
# '../data/R_out/Phyloseq/P.Controls.16S.species.Count.RDS       
#                 --- count data (sequence reads)
# '../data/R_out/Phyloseq/P.Soils.16S.species.Count.RDS          
#                 --- count data (sequence reads)
# '../data/R_out/Phyloseq/P.Soils.16S.species.RelAbund.RDS       
#                 --- rel abund (not normalized by 16S)
# '../data/R_out/Phyloseq/P.Soils.GIBBs.RelAbund.RDS             
#                 --- rel abund (normalized by 16S & functional gene copy numbers)
# '../data/R_out/Phyloseq/P.Soils.16S.species.Total.RDS'
#                 --- PLFA pmol g-1 soil
#                 --- DNA copies g-1 soil (not normalized by 16S)
# '../data/R_out/Phyloseq/P.Soils.16S.GIBBs.Total.RDS'
#                 --- PLFA pmol g-1 soil
#                 --- DNA copies g-1 soil (normalized by 16S & func gene copies)


```{r}
library('daymetr')
library('MASS')
library('readxl')
library('phyloseq')
library('tidyverse')
library('writexl')
library('vegan')
library('matrixStats')
library('lavaan')
library('semPlot')
library('caret')

source('SHAI_Functions.R')
source("myFunctions.R")

geom_text_repel2 <- function(...) {
    layer <- ggrepel::geom_text_repel(...)
    layer$ggrepel <- TRUE
    class(layer) <- c("ggrepel", class(layer))
    return(layer)
}

ggplot_add.ggrepel <- function(object, plot, object_name) {
    if (any(do.call(c, lapply(plot$layer, function(x) x$ggrepel)))) {
        warning(
            "There is more than one ggrepel layers. ",
            "This may cause overlap of labels"
        )
    }
   # Optionally, one may modify `object` here.
    NextMethod("ggplot_add")
}
```


### Run this section to import excel file
```{r} 
df <- as.data.frame(read_excel("../data/raw_data/myPhyloDB/SHAI.Meta.22March2023.xlsm", sheet="Final", skip=1))
row.names(df) <- df$`PLFA_ID`
df
```


### Calclulate PLFA indicator groups
```{r plfa}
# Extract just PLFA data
df.plfa <- df[,c(96:207)]
names(df.plfa)

# Calculate PLFA indicators using custom function
df.plfa.markers <- plfa_indicators(df.plfa, markers='PLFA_markers.xlsx') # indicators can be changed by modifying this file
df.plfa.markers
```


### Run this section to get Soil Health calculations
```{r SHI}
# Extract just meta data
df.meta <- df[,c(1:60)]
names(df.meta)

# plant richness & SHMS 
df.meta.shi <- soil_health_index(df.meta,
                                 func_groups='Plant_functional_groups.xlsx',
                                 till_groups='Tillage_scores.xlsx', #scores can be changed by modifying this file
                                 mixture_weight='6', #richness for mixture (Other category)
                                 mixture_func_weight='3' #functional richness for mixture (Other category)
                                 )
df.meta.shi
```


## Get DAYMET data for each location
```{r daymet}
### Run this section to get Daymet data
#########################################################################
# write coordinates file
coords <- data.frame(sample=df$PLFA_ID, latitude=round(df$Latitude, 3), longitude=round(df$Longitude, 3))
write.table(coords, file="../data/R_out/Tabular/Coords.csv", row.names=FALSE, sep=",")

# get data
data <- download_daymet_batch(
  file_location = '../data/R_out/Tabular/Coords.csv',
  start = 2008,
  end = 2017,
  internal = TRUE,
  simplify=TRUE)

#select the years wanted
wanted <- data %>%
  filter(year >= 2008)
  
#calculate mean annual temp from max and min and years selected
annual_temp <- wanted %>%
  group_by(site, year) %>%
  filter(measurement == "tmax..deg.c." | measurement == "tmin..deg.c.") %>%
  summarize(MAT = mean(as.numeric(value)))

mean_annual_temp <- annual_temp %>%
  group_by(site) %>%
  summarize(MAT = mean(MAT))

annual_ppt <- wanted %>%
  group_by(site, year) %>%
  filter(measurement == 'prcp..mm.day.') %>%
  summarize(PPT = sum(as.numeric(value)))

mean_annual_ppt <- annual_ppt %>%
  group_by(site) %>%
  summarize(PPT = mean(PPT))
  
annual_srad <- wanted %>%
  group_by(site, year) %>%
  filter(measurement == 'srad..W.m.2.') %>%
  summarize(SRAD = sum(as.numeric(value)))

mean_annual_srad <- annual_srad %>%
  group_by(site) %>%
  summarize(SRAD = mean(SRAD))

annual_swe <- wanted %>%
  group_by(site, year) %>%
  filter(measurement == 'swe..kg.m.2.') %>%
  summarize(SWE = sum(as.numeric(value)))

mean_annual_swe <- annual_swe %>%
  group_by(site) %>%
  summarize(SWE = mean(SWE))

annual_dayl <- wanted %>%
  group_by(site, year) %>%
  filter(measurement == 'dayl..s.') %>%
  summarize(DLEN = sum(as.numeric(value)))

mean_annual_dayl <- annual_dayl %>%
  group_by(site) %>%
  summarize(DLEN = mean(DLEN))

svp <- wanted %>%
  group_by(site, year, yday) %>%
  filter(measurement == "tmax..deg.c." | measurement == "tmin..deg.c.") %>%
  summarize(MAT = mean(as.numeric(value))) %>%
  mutate(SVP = 610.7*10^(7.5*MAT/(237.3+MAT))) 

vp <- wanted %>%
  group_by(site, year, yday) %>%
  filter(measurement == 'vp..Pa.')

vpd <- svp %>%
  full_join(vp) %>%
  mutate(VPD = as.numeric(SVP) - as.numeric(value))
  
annual_vpd <- vpd %>%
  group_by(site, year) %>%
  summarize(VPD = sum(VPD))

mean_annual_vpd <- annual_vpd %>%
  group_by(site) %>%
  summarize(VPD = mean(VPD))

df.mat <- data.frame(MAT=mean_annual_temp$MAT, PPT=mean_annual_ppt$PPT, VPD=mean_annual_vpd$VPD, SRAD=mean_annual_srad$SRAD, SWE=mean_annual_swe$SWE, DLEN=mean_annual_dayl$DLEN)
row.names(df.mat) <- mean_annual_temp$site
```


## write the final dataset
```{r}
#########################################################################
df.rating <- merge(df, df.plfa.markers, by='row.names', all=TRUE)
df.rating <- select(df.rating,-c(Row.names, Index))
row.names(df.rating) <- df.rating$PLFA_ID

df.rating <- merge(df.rating, df.meta.shi, by='row.names', all=TRUE)
df.rating <- select(df.rating,-c(Row.names))
row.names(df.rating) <- df.rating$PLFA_ID

#df.rating <- merge(df.rating, df.mat, by='row.names', all=TRUE)
#df.rating <- select(df.rating,-c(Row.names))
#row.names(df.rating) <- df.rating$PLFA_ID

write_xlsx(df.rating, '../data/R_out/Tabular/SHAI.All_data.Final.xlsx')
#########################################################################
```


## Convert myPhyloDB data to new Phyloseq object 
# All data has been updated (Feb 2022) and re-classified with the new EMU platform.
```{r}
### load NRCS data and wrangle into phyloseq
##################################################################################
## get the meta data previously imported into myPhyloDB
biom <- import_biom('../data/raw_data/myPhyloDB/NRCS.myphylodb.biom')
SAMP <- sample_data(biom)
sample_names(SAMP) <- SAMP$sample_name

## get the new taxonomy data from emu
tax <- read.csv('../data/raw_data/emu/NRCS.taxonomy.csv', sep='\t', header=F)
names(tax) <- c('tax_id', 'Taxonomy')
tax <- tax %>%
  separate(Taxonomy, c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species'), sep=';')
tax[is.na(tax)] <- ""
tax <- tax.clean(tax)
TAX <- tax_table(as.matrix(tax[,c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'tax_id')]))
taxNames <- taxa_names(TAX)

## get the abundance data from myPhyloDB
otu <- read.csv('../data/raw_data/myPhyloDB/NRCS.shared.txt', sep='\t') %>%
  select(-group)
OTU <- otu_table(otu, taxa_are_rows=T)
taxa_names(OTU) <- taxNames

## create initial phyloseq object (ASV's)
P <- phyloseq(SAMP, OTU, TAX)
saveRDS(P, file='../data/R_out/Phyloseq/P.NRCS.16S.asv.RDS')

# Combine duplicate tax_ids
P.taxid <- tax_glom(P, taxrank='tax_id')
taxa_names(P.taxid) <- tax_table(P.taxid)[,'tax_id']
saveRDS(P.taxid, file='../data/R_out/Phyloseq/P.NRCS.16S.taxid.RDS')

# Combine by Species (will be used for merging)
P.species <- tax_glom(P.taxid, taxrank='Species')
taxa_names(P.species) <- tax_table(P.species)[,'Species']
saveRDS(P.species, file='../data/R_out/Phyloseq/P.NRCS.16S.species.RDS')
##################################################################################


### load Hops.ARS data and wrangle into phyloseq
##################################################################################
## get the meta data previously imported into myPhyloDB
biom <- import_biom('../data/raw_data/myPhyloDB/Hops.ARS.myphylodb.biom')
SAMP <- sample_data(biom)
sample_names(SAMP) <- SAMP$sample_name

## get the taxonomy data from emu
tax <- read.csv('../data/raw_data/emu/Hops.ARS.taxonomy.csv', sep='\t', header=F)
names(tax) <- c('tax_id', 'Taxonomy')
tax <- tax %>%
  separate(Taxonomy, c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species'), sep=';')
tax[is.na(tax)] <- ""
tax <- tax.clean(tax)
TAX <- tax_table(as.matrix(tax[,c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'tax_id')]))
taxNames <- taxa_names(TAX)

## get the abundance data from myPhyloDB
otu <- read.csv('../data/raw_data/myPhyloDB/Hops.ARS.shared.txt', sep='\t') %>%
  select(-group)
OTU <- otu_table(otu, taxa_are_rows=T)
taxa_names(OTU) <- taxNames

## create initial phyloseq object (ASV's)
P <- phyloseq(SAMP, OTU, TAX)
saveRDS(P, file='../data/R_out/Phyloseq/P.Hops.ARS.16S.asv.RDS')

# Combine duplicate tax_ids
P.taxid <- tax_glom(P, taxrank='tax_id')
taxa_names(P.taxid) <- tax_table(P.taxid)[,'tax_id']
saveRDS(P.taxid, file='../data/R_out/Phyloseq/P.Hops.ARS.16S.taxid.RDS')

# Combine by Species (will be used for merging)
P.species <- tax_glom(P.taxid, taxrank='Species')
taxa_names(P.species) <- tax_table(P.species)[,'Species']
saveRDS(P.species, file='../data/R_out/Phyloseq/P.Hops.ARS.16S.species.RDS')
##################################################################################


### load Hops.2018 data and wrangle into phyloseq
##################################################################################
## get the meta data previously imported into myPhyloDB
biom <- import_biom('../data/raw_data/myPhyloDB/Hops.2018.myphylodb.biom')
SAMP <- sample_data(biom)
sample_names(SAMP) <- gsub('2018.', '', SAMP$sample_name)

## get the taxonomy data from emu
tax <- read.csv('../data/raw_data/emu/Hops.2018.taxonomy.csv', sep='\t', header=F)
names(tax) <- c('tax_id', 'Taxonomy')
tax <- tax %>%
  separate(Taxonomy, c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species'), sep=';')
tax[is.na(tax)] <- ""
tax <- tax.clean(tax)
TAX <- tax_table(as.matrix(tax[,c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'tax_id')]))
taxNames <- taxa_names(TAX)

## get the abundance data from myPhyloDB
otu <- read.csv('../data/raw_data/myPhyloDB/Hops.2018.shared.txt', sep='\t') %>%
  select(-group)
OTU <- otu_table(otu, taxa_are_rows=T)
taxa_names(OTU) <- taxNames

## create initial phyloseq object (ASV's)
P <- phyloseq(SAMP, OTU, TAX)
saveRDS(P, file='../data/R_out/Phyloseq/P.Hops.2018.16S.asv.RDS')

# Combine duplicate tax_ids
P.taxid <- tax_glom(P, taxrank='tax_id')
taxa_names(P.taxid) <- tax_table(P.taxid)[,'tax_id']
saveRDS(P.taxid, file='../data/R_out/Phyloseq/P.Hops.2018.16S.taxid.RDS')

# Combine by Species (will be used for merging)
P.species <- tax_glom(P.taxid, taxrank='Species')
taxa_names(P.species) <- tax_table(P.species)[,'Species']
saveRDS(P.species, file='../data/R_out/Phyloseq/P.Hops.2018.16S.species.RDS')
##################################################################################


### load OR.PMC data and wrangle into phyloseq
##################################################################################
## get the meta data previously imported into myPhyloDB
biom <- import_biom('../data/raw_data/myPhyloDB/OR.PMC.myphylodb.biom')
SAMP <- sample_data(biom)
sample_names(SAMP) <- SAMP$sample_name

## get the taxonomy data from emu
tax <- read.csv('../data/raw_data/emu/OR.PMC.taxonomy.csv', sep='\t', header=F)
names(tax) <- c('tax_id', 'Taxonomy')
tax <- tax %>%
  separate(Taxonomy, c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species'), sep=';')
tax[is.na(tax)] <- ""
tax <- tax.clean(tax)
TAX <- tax_table(as.matrix(tax[,c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'tax_id')]))
taxNames <- taxa_names(TAX)

## get the abundance data from myPhyloDB
otu <- read.csv('../data/raw_data/myPhyloDB/OR.PMC.shared.txt', sep='\t', check.names=F) %>%
  select(-group)
OTU <- otu_table(otu, taxa_are_rows=T)
taxa_names(OTU) <- taxNames

## create initial phyloseq object (ASV's)
P <- phyloseq(SAMP, OTU, TAX)
saveRDS(P, file='../data/R_out/Phyloseq/P.OR.PMC.16S.asv.RDS')

# Combine duplicate tax_ids
P.taxid <- tax_glom(P, taxrank='tax_id')
taxa_names(P.taxid) <- tax_table(P.taxid)[,'tax_id']
saveRDS(P.taxid, file='../data/R_out/Phyloseq/P.OR.PMC.16S.taxid.RDS')

# Combine by Species (will be used for merging)
P.species <- tax_glom(P.taxid, taxrank='Species')
taxa_names(P.species) <- tax_table(P.species)[,'Species']
saveRDS(P.species, file='../data/R_out/Phyloseq/P.OR.PMC.16S.species.RDS')
##################################################################################


### load Rangeland data and wrangle into phyloseq
##################################################################################
## get the meta data previously imported into myPhyloDB
biom <- import_biom('../data/raw_data/myPhyloDB/Rangeland.myphylodb.biom')
SAMP <- sample_data(biom)
sample_names(SAMP) <- SAMP$sample_name

## get the taxonomy data from emu
tax <- read.csv('../data/raw_data/emu/Rangeland.taxonomy.csv', sep='\t', header=F)
names(tax) <- c('tax_id', 'Taxonomy')
tax <- tax %>%
  separate(Taxonomy, c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species'), sep=';')
tax[is.na(tax)] <- ""
tax <- tax.clean(tax)
TAX <- tax_table(as.matrix(tax[,c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'tax_id')]))
taxNames <- taxa_names(TAX)

## get the abundance data from myPhyloDB
otu <- read.csv('../data/raw_data/myPhyloDB/Rangeland.shared.txt', sep='\t') %>%
  select(-group)
OTU <- otu_table(otu, taxa_are_rows=T)
taxa_names(OTU) <- taxNames

## create initial phyloseq object (ASV's)
P <- phyloseq(SAMP, OTU, TAX)
saveRDS(P, file='../data/R_out/Phyloseq/P.Rangeland.16S.asv.RDS')

# Combine duplicate tax_ids
P.taxid <- tax_glom(P, taxrank='tax_id')
taxa_names(P.taxid) <- tax_table(P.taxid)[,'tax_id']
saveRDS(P.taxid, file='../data/R_out/Phyloseq/P.Rangeland.16S.taxid.RDS')

# Combine by Species (will be used for merging)
P.species <- tax_glom(P.taxid, taxrank='Species')
taxa_names(P.species) <- tax_table(P.species)[,'Species']
saveRDS(P.species, file='../data/R_out/Phyloseq/P.Rangeland.16S.species.RDS')
##################################################################################


### combine all phyloseq objects
##################################################################################
P1 <- readRDS('../data/R_out/Phyloseq/P.Hops.ARS.16S.species.RDS')
P2 <- readRDS('../data/R_out/Phyloseq/P.Hops.2018.16S.species.RDS')
P3 <- readRDS('../data/R_out/Phyloseq/P.NRCS.16S.species.RDS')
P4 <- readRDS('../data/R_out/Phyloseq/P.OR.PMC.16S.species.RDS')
P5 <- readRDS('../data/R_out/Phyloseq/P.Rangeland.16S.species.RDS')

P <- merge_phyloseq(P1, P2, P3, P4, P5)

# Add SHAI codes to data
codes <- read.csv('../data/raw_data/myPhyloDB/shai_code.csv')
row.names(codes) <- codes$sample_name
codes <- codes[,2:3]
SAMP <- merge(sample_data(P), codes, by=0, all=T)
row.names(SAMP) <- SAMP$Row.names
sample_data(P) <- SAMP

# Get controls
P.controls <- subset_samples(P, Control != "")
saveRDS(P.controls, file='../data/R_out/Phyloseq/P.Controls.16S.species.Count.RDS')

# Merge replicate samples (Hops, PMC)
P.soils <- subset_samples(P, SHAI.code != "")
P.merge <- merge_samples(P.soils, 'SHAI.code')

# replace sample data with newly calculated data
dat <- read_excel('../data/R_out/Tabular/SHAI.All_data.Final.xlsx')
SAMP <- sample_data(dat)
sample_names(SAMP) <- SAMP$PLFA_ID
sample_data(P.merge) <- SAMP
saveRDS(P.merge, file='../data/R_out/Phyloseq/P.Soils.16S.species.Count.RDS')

P.rel <- transform_sample_counts(P.merge, function (x) x / sum(x))
saveRDS(P.rel, file='../data/R_out/Phyloseq/P.Soils.16S.species.RelAbund.RDS')

# convert sequence count data to total abundance (genomes / g soil)
P.total <- P.rel
for (n in 1:nsamples(P.total)) {
  otu_table(P.total)[,n] <- otu_table(P.total)[,n] * sample_data(P.rel)$qPCR...16S.copies...g.DW.soil.[n]
}
# convert PLFA data to total abundance (pmol / g soil)
sample_data(P.total)[,c(95:206, 211:217)] <- sample_data(P.total)[,c(95:206, 211:217)] * sample_data(P.rel)$Total.PLFA....pmol.g.soil.

saveRDS(P.total, file='../data/R_out/Phyloseq/P.Soils.16S.species.Total.RDS')
################################################################################
```


## Create phyloseq object with GIBBs data 
# PICRUSt2 was previously run using 'SHAI.miSeq.PICRUSt2.Rmd'
# just adding meta data
```{r}
### load NRCS data and wrangle into phyloseq
##################################################################################
P.GIBBs <- readRDS('../data/picrust2/NRCS.GIBBs.phyloseq.RDS')
P.tax <- readRDS('../data/R_out/Phyloseq/P.NRCS.16S.species.RDS')
sample_data(P.GIBBs) <- sample_data(P.tax)
saveRDS(P.GIBBs, file='../data/R_out/Phyloseq/P.NRCS.GIBBs.RDS')
##################################################################################


### load Hops.ARS data and wrangle into phyloseq
##################################################################################
P.GIBBs <- readRDS('../data/picrust2/Hops.ARS.GIBBs.phyloseq.RDS')
P.tax <- readRDS('../data/R_out/Phyloseq/P.Hops.ARS.16S.species.RDS')
sample_data(P.GIBBs) <- sample_data(P.tax)
saveRDS(P.GIBBs, file='../data/R_out/Phyloseq/P.Hops.ARS.GIBBs.RDS')
##################################################################################


### load Hops.2018 data and wrangle into phyloseq
##################################################################################
P.GIBBs <- readRDS('../data/picrust2/Hops.2018.GIBBs.phyloseq.RDS')
sample_names(P.GIBBs) <- gsub('_', '.s', sample_names(P.GIBBs))
P.tax <- readRDS('../data/R_out/Phyloseq/P.Hops.2018.16S.species.RDS')
sample_data(P.GIBBs) <- sample_data(P.tax)
saveRDS(P.GIBBs, file='../data/R_out/Phyloseq/P.Hops.2018.GIBBs.RDS')
##################################################################################


### load OR.PMC data and wrangle into phyloseq
##################################################################################
P.GIBBs <- readRDS('../data/picrust2/OR.PMC.GIBBs.phyloseq.RDS')
sample_names(P.GIBBs) <- gsub('^X', '', sample_names(P.GIBBs))
P.tax <- readRDS('../data/R_out/Phyloseq/P.OR.PMC.16S.species.RDS')
sample_data(P.GIBBs) <- sample_data(P.tax)
saveRDS(P.GIBBs, file='../data/R_out/Phyloseq/P.OR.PMC.GIBBs.RDS')
##################################################################################


### load Rangeland data and wrangle into phyloseq
##################################################################################
P.GIBBs <- readRDS('../data/picrust2/Rangeland.GIBBs.phyloseq.RDS')
P.tax <- readRDS('../data/R_out/Phyloseq/P.Rangeland.16S.species.RDS')
sample_data(P.GIBBs) <- sample_data(P.tax)
saveRDS(P.GIBBs, file='../data/R_out/Phyloseq/P.Rangeland.GIBBs.RDS')
##################################################################################


### combine all phyloseq objects
##################################################################################
P1 <- readRDS('../data/R_out/Phyloseq/P.Hops.ARS.GIBBs.RDS')
P2 <- readRDS('../data/R_out/Phyloseq/P.Hops.2018.GIBBs.RDS')
P3 <- readRDS('../data/R_out/Phyloseq/P.NRCS.GIBBs.RDS')
P4 <- readRDS('../data/R_out/Phyloseq/P.OR.PMC.GIBBs.RDS')
P5 <- readRDS('../data/R_out/Phyloseq/P.Rangeland.GIBBs.RDS')

P <- merge_phyloseq(P1, P2, P3, P4, P5)

# Add SHAI codes to data
codes <- read.csv('../data/raw_data/myPhyloDB/shai_code.csv')
row.names(codes) <- codes$sample_name
codes <- codes[,2:3]
SAMP <- merge(sample_data(P), codes, by=0, all=T)
row.names(SAMP) <- SAMP$Row.names
sample_data(P) <- SAMP

# Merge replicate samples (Hops, PMC)
P.soils <- subset_samples(P, SHAI.code != "")
P.merge <- merge_samples(P.soils, 'SHAI.code', fun=sum)

# replace sample data with newly calculated data
dat <- read_excel('../data/R_out/Tabular/SHAI.All_data.Final.xlsx')
SAMP <- sample_data(dat)
sample_names(SAMP) <- SAMP$PLFA_ID
sample_data(P.merge) <- SAMP
saveRDS(P.merge, file='../data/R_out/Phyloseq/P.Soils.GIBBs.RelAbund.RDS')

# convert sequence count data to total abundance (genomes / g soil)
P.total <- P.merge
for (n in 1:nsamples(P.total)) {
  otu_table(P.total)[n,] <- otu_table(P.total)[n,] * sample_data(P.merge)$qPCR...16S.copies...g.DW.soil.[n]
}

# convert PLFA data to toal abundance (pmol / g soil)
sample_data(P.total)[,c(95:206, 211:217)] <- sample_data(P.total)[,c(95:206, 211:217)] * sample_data(P.merge)$Total.PLFA....pmol.g.soil.

saveRDS(P.total, file='../data/R_out/Phyloseq/P.Soils.GIBBs.Total.RDS')
##################################################################################
```


## write the final Relative Abundance dataset with sequence data to Excel
```{r}
#########################################################################
# Calculate alpha diversity indices
P.soils.taxa <- readRDS('../data/R_out/Phyloseq/P.Soils.16S.species.Count.RDS')
rich <- estimate_richness(P.soils.taxa, measures=c("Observed", "Chao1", "ACE", "Shannon", "InvSimpson"))
names(rich) <- c('Sobs', 'Chao1', 'SE.Chao1', 'ACE', 'SE.ACE', 'Shannon', 'InvSimpson')
rich$PLFA_ID <- row.names(rich)
rich$PLFA_ID <- gsub('\\.', '-', rich$PLFA_ID)

# Calculate phyla relative abundances
P.soils.taxa <- readRDS('../data/R_out/Phyloseq/P.Soils.16S.species.RelAbund.RDS')
P.glom <- tax_glom(P.soils.taxa, taxrank="Phylum")
phyla <- data.frame(otu_table(P.glom))
names(phyla) <- tax_table(P.glom)[,'Phylum']
phyla$PLFA_ID <- row.names(phyla)

# GIBBs relative abundances
P.soils.GIBBs <- readRDS('../data/R_out/Phyloseq/P.Soils.GIBBs.RelAbund.RDS')
GIBBs <- data.frame(otu_table(P.soils.GIBBs))
names(GIBBs) <- tax_table(P.soils.GIBBs)[,'Short']
GIBBs$PLFA_ID <- row.names(GIBBs)

meta <- read_excel('../data/R_out/Tabular/SHAI.All_data.Final.xlsx')

final <- meta %>%
  left_join(rich, by='PLFA_ID')
final <- final %>%
  left_join(phyla, by='PLFA_ID')
final <- final %>%
  left_join(GIBBs, by='PLFA_ID')

write_xlsx(final, '../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.EC.RelAbund.xlsx')
#########################################################################
```

## write the final Total Abundance dataset  to Excel
```{r}
#########################################################################
ra <- read_excel('../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.EC.RelAbund.xlsx')

tot <- ra %>%
  mutate(across(c(95:206, 211:217), ~ .x * `Total PLFA \r\n(pmol/g soil)`)) %>%
  mutate(across(c(246:341), ~ .x * `qPCR\r\n(16S copies / g DW soil)`))

write_xlsx(tot, '../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.EC.Total.xlsx')
#########################################################################
```


## PCoA
```{r}
ra <- read_excel('../data/R_out/Tabular/SHAI.All_data.Final.PLFA.Phyla.KO.RelAbund.xlsx')
temp <- ra[,c(61:64, 70, 72, 74, 78, 80, 82, 84, 86, 88:91, 207, 209, 246:341)]
temp <- temp[complete.cases(temp),]
temp <- temp %>%
   filter(!row_number() %in% c(421,422))

preProcValues <- preProcess(temp, method='range')
scale <- predict(preProcValues, temp)

cash <- scale[,c('clay', 'silt', 'water_cap', 'agg_stab', 'SOM', 'ace', 'resp', 'activeC', 'ph', 'p', 'k', 'mg', 'fe', 'mn', 'zn', 'Total PLFA \r\r\n(pmol/g soil)', 'qPCR\r\r\n(16S copies / g DW soil)')]
taxa <- scale[,c(19:24,26:66)]
gibbs <- scale[,c(67:114)]

#gibbs <- gibbs %>%
#  mutate(biocontrol = scale(rowMedians(as.matrix(select(., c(hcnA, budA, budC, E3.2.1.6, E3.2.1.14, phlD, fenA, srfAA, rifM, phzE)))))) %>%
#  mutate(stress = scale(rowMedians(as.matrix(select(., c(ipdC, acdS)))))) %>%
#  mutate(siderophore = scale(rowMedians(as.matrix(select(., c(mbtI, entA, pchB)))))) %>%
#  mutate(C_decomp = scale(rowMedians(as.matrix(select(., c(bglB, bglX, E3.2.1.21)))))) %>%
#  mutate(N_decomp = scale(rowMedians(as.matrix(select(., c(amiE, ureC)))))) %>%
#  mutate(P_decomp = scale(rowMedians(as.matrix(select(., c(phoA, phoD, PHO, appA, phoN)))))) %>%
#  mutate(N_fix = scale(rowMedians(as.matrix(select(., c(nifH, nifD, nifK)))))) %>%
#  mutate(Nitrif = scale(rowMedians(as.matrix(select(., c(`pmoA-amoA`, hao)))))) %>%
#  mutate(Denitrif = scale(rowMedians(as.matrix(select(., c(narG, narH, nirK, nirS, norB, norC, nosZ)))))) %>%
#  mutate(EPS = scale(rowMedians(as.matrix(select(., c(EpsA, EpsB, EpsC, EpsD))))))


## PCoA
ord <- rda(cash)

species <-as.data.frame(scores(ord, display='species', choices=c(1:2)))
names(species) <- c("Axis1", "Axis2")
species$lab <- row.names(species)

fit <- envfit(ord ~ ., taxa)
fit
biplot <- select.envfit(fit, 0.05)
biplot <- as.data.frame(scores(biplot, display='vectors', choices=c(1:2)))
biplot <- na.omit(biplot)
names(biplot) <- c("Axis1", "Axis2")
biplot$lab <- row.names(biplot)
biplot$colors <- 'red'

### get eigenvalues and proportion explained
ev <- eigenvals(ord)
Axis1_exp = round(ev[1] / sum(ev) * 100, 1)
Axis2_exp = round(ev[2] / sum(ev) * 100, 1)

### get PC axes
axes <- as.data.frame(scores(ord, display="sites", choices=c(1:2)))
names(axes) <- c("Axis1", "Axis2")


p <- ggplot(data=axes, aes(x=Axis1, y=Axis2)) +
  geom_point(size=6, color='grey') +
  labs(x=paste0('RDA1 (', Axis1_exp, '%)'), 
       y=paste0('RDA2 (', Axis2_exp, '%)')) +
  geom_segment(data = species, inherit.aes=F,
             aes(x=0, xend=Axis1, y=0, yend=Axis2), color='black',
             arrow = arrow(length = unit(0.1, "cm"))) +
  geom_text_repel2(data = species, inherit.aes=F,
                           aes(x=Axis1, y=Axis2, label=lab),
                           segment.size=0.25, segment.linetype=1, segment.color='grey',
                           color='black', size=3, direction='both', max.overlaps=50) +
  geom_segment(data = biplot, inherit.aes=F,
             aes(x=0, xend=Axis1, y=0, yend=Axis2), color='red',
             arrow = arrow(length = unit(0.1, "cm"))) +
  geom_text_repel2(data = biplot, inherit.aes=F,
                           aes(x=Axis1, y=Axis2, label=lab),
                           segment.size=0.25, segment.linetype=1, segment.color='grey',
                           color='red', size=3, direction='both', max.overlaps=50)
p



p <- ggplot(gibbs[,c(49:58)], aes(x=biocontrol, y=Nitrif)) +
  geom_point()


model <- '
    mgt =~ all_roots_12mosbeforecol_bin + all_total_func_bin + till_bin
    SH1 =~ resp_z + copiespergsoil16_z + Total_PLFA_z
    ClimateZ ~ mgt 
    texture_class ~ mgt
    SH1 + ace_z + agg_stab_z + activeC_z + water_cap_z + SOM_z ~ ClimateZ + texture_class + mgt
'
fit <- sem(model, data=final)

inspect(fit, "cor.lv")

summary(fit, rsquare=T, standardized=T, fit.measures=T)
semPaths(fit, "std", bg="white", edge.label.cex=1.5, thresholds=F)


XFormula <- ~ Trt + clay + silt + water_cap + agg_stab + SOM + ace + resp + activeC

studyDesign <- data.frame(plot = as.factor(meta$Transect))
rL <- HmscRandomLevel(units = levels(studyDesign$plot))

mod = Hmsc(Y = comm, XData = meta, XFormula = XFormula, 
           studyDesign = studyDesign,
           ranLevels = list(plot = rL))

samples <- nrow(d)
nChains = 2
test.run = TRUE
if (test.run){
  thin = 1
  samples = 10
  transient = ceiling(thin*samples*.5)
  verbose = 50
  hmsc_file <- "hmsc_probit_test_subplot.Rda"
} else {
  thin = 10
  samples = 1000
  transient = ceiling(thin*samples*.5)
  verbose = 50
  hmsc_file <- "hmsc_probit_subplot.Rda"
}

m = sampleMcmc(mod, thin = thin, samples = samples, transient = transient,
               nChains = nChains, nParallel = nChains, verbose = verbose)

# check convergence diagnostics
mpost = convertToCodaObject(m)
par(mfrow=c(2,2))
hist(effectiveSize(mpost$Beta), main="ess(beta)")
hist(gelman.diag(mpost$Beta, multivariate=FALSE)$psrf, main="psrf(beta)")
hist(effectiveSize(mpost$Omega[[1]]), main="ess(omega)")
hist(gelman.diag(mpost$Omega[[1]], multivariate=FALSE)$psrf, main="psrf(omega)")

postBeta = getPostEstimate(m, parName="Beta")
plotBeta(m, post=postBeta, param="Support", supportLevel = 0.95)


```

## Hmsc
```{r}
library(Hmsc)
library(corrplot)
require(snow)
library(ggpubr)
library(ggthemes)
library(ggtext)
library(caret)

load("../data/raw_data/myPhyloDB/df_filt.Rdata")

mgt <- df_filt[,c('all_roots_12mosbeforecol_bin', 'all_annual_func_rich_bin', 'till_bin')]
names(mgt) <- c('root', 'rich', 'dist')
cash <- df_filt[,c('water_cap', 'agg_stab', 'SOM', 'ace', 'resp', 'activeC', 'p', 'k', 'mg', 'fe', 'zn', 'mn', 'ph')]
taxa <- df_filt[,c(255:260, 262:302)]
gibbs <- df_filt[,c(303:344, 346:350)]
other <- df_filt[,c('ClimateZ', 'texture_class')]

preProcValues <- preProcess(cash, method='range')
cash.scale <- predict(preProcValues, cash)

preProcValues <- preProcess(gibbs, method='range')
gibbs.scale <- predict(preProcValues, gibbs)

preProcValues <- preProcess(taxa, method='range')
taxa.scale <- predict(preProcValues, taxa)

gene <- gibbs %>%
  mutate(biocontrol = rowMeans(dplyr::select(., c(hcnA, budA, budC, E3.2.1.6, E3.2.1.14, phlD, fenA, srfAA, rifM, phzE)))) %>%
  mutate(stress = rowMeans(dplyr::select(., c(ipdC, acdS)))) %>%
  mutate(siderophore = rowMeans(dplyr::select(., c(mbtI, entA, pchB)))) %>%
  mutate(C_decomp = rowMeans(dplyr::select(., c(bglB, bglX, E3.2.1.21)))) %>%
  mutate(N_decomp = rowMeans(dplyr::select(., c(amiE, ureC)))) %>%
  mutate(P_decomp = rowMeans(dplyr::select(., c(phoA, phoD, PHO, appA, phoN)))) %>%
  mutate(N_fix = rowMeans(dplyr::select(., c(nifH, nifD, nifK)))) %>%
  mutate(Nitrif = rowMeans(dplyr::select(., c(`pmoA-amoA`, hao)))) %>%
  mutate(Denitrif = rowMeans(dplyr::select(., c(narG, narH, nirK, nirS, norB, norC, nosZ)))) %>%
  mutate(EPS = rowMeans(dplyr::select(., c(EpsA, EpsB, EpsC, EpsD))))

P.soils.taxa <- readRDS('../data/R_out/Phyloseq/P.Soils.16S.species.RelAbund.RDS')
P.sub <- subset_samples(P.soils.taxa, PLFA_ID %in% df_filt$PLFA_ID)
#P.glom <- tax_glom(P.sub, taxrank="Genus")

wh0 = genefilter_sample(P.sub, filterfun_sample(function(x) x > 0), A=362*0.50) #91 species
P.sub.sub = prune_taxa(wh0, P.sub) ##Remove the selected OTUs

TopNOTUs = names(sort(taxa_sums(P.glom), TRUE)[1:50])
P.50 = prune_taxa(TopNOTUs, P.glom)
species <- data.frame(otu_table(P.50))
names(species) <- tax_table(P.50)[,'species']

final <- cbind(other, mgt, species)
final$ClimateZ <- factor(final$ClimateZ)
final$texture_class <- factor(final$texture_class)
final$root <- factor(final$root)
final$rich <- factor(final$rich)
final$dist <- factor(final$dist)

n <- nrow(cash)
studyDesign <- data.frame(sample = as.factor(1:n))
rL <- HmscRandomLevel(units = studyDesign$sample)

mod = Hmsc(Y = cash, XData = final, XFormula = ~ ., 
           distr="poisson",
           studyDesign = studyDesign,
           ranLevels = list(sample = rL))

nChains = 5
test.run = FALSE
if (test.run){
  thin = 1
  samples = 10
  transient = 5
  verbose = 50
} else {
  thin = 10
  samples = 1000
  transient = 500
  verbose = 50
}

m = sampleMcmc(mod, thin = thin, samples = samples, transient = transient,
               nChains = nChains, nParallel = nChains, verbose = verbose)

# getting the posteriors
mpost <- convertToCodaObject(m)
preds = computePredictedValues(m)
MF = evaluateModelFit(hM=m, predY=preds)
VP <- computeVariancePartitioning(m)

# model convergence, diagnostics
ess.beta <- effectiveSize(mpost$Beta) %>%
  as_tibble() %>% 
  dplyr::mutate(variable = "Effective Sample Size")
psrf.beta <- gelman.diag(mpost$Beta, multivariate=FALSE)$psrf%>%
  as_tibble() %>% 
  dplyr::rename(value = `Point est.`) %>%
  dplyr::mutate(variable = "Gelman Diagnostic")

diag_all <- bind_rows(ess.beta, psrf.beta) %>%
  ggplot(aes(x=value)) +
  geom_histogram() +
  theme_classic() +
  facet_wrap(~variable, scales="free") +
  ggtitle("Convergence Diagnostics")+
  theme(plot.background = element_rect( color ="black"),
        axis.title.x = element_blank(),
        plot.title = element_text(hjust = 1, face = "bold"))
diag_all

ggsave(diag_all, filename="cash.species.gelman_ess.pdf", 
       width=8, height=6, bg="white")

hist(MF$SR2, xlim = c(0,1), main=paste0("Mean = ", round(mean(MF$SR2),2)))

# explanatory power
#MF$TjurR2%>% mean(na.rm=T)
#ggarrange(
#  ggplot(as.data.frame(MF),aes(x=(RMSE))) + geom_histogram(),
#  ggplot(as.data.frame(MF),aes(x=(TjurR2))) + geom_histogram(),
#  ggplot(as.data.frame(MF),aes(x=(AUC))) + geom_histogram())

# variance partitioning
mf_df <- as.data.frame(MF) %>%
  dplyr::mutate(Species = colnames(m$Y)) %>%
  dplyr::select(-O.AUC, -O.TjurR2) %>%
  tidyr::pivot_longer(cols = names(.)[1:5], 
               names_to = 'variable',
               values_to = 'value') %>%
  mutate(category = case_when(
    variable %in% c("SR2", "C.SR2", "R2") ~ "R2",
    variable %in% c("RMSE", "C.RMSE", "O.RMSE") ~ "RMSE")) %>%
  group_by(variable) %>%
  mutate(avg = mean(value) %>% round(2)) %>%
  ungroup() %>%
  mutate(variable = paste0(variable, ": (", avg, ")"))

p_fit <- ggplot(mf_df, aes(x=value, y=Species, fill = variable)) +
  geom_bar(position = 'dodge', stat = 'identity') +
  facet_wrap(~category, scales = "free") +
  scale_fill_brewer(palette = "Set1", name = "Fit Measure\n(Avg)") +
  labs(title = "Fit Measures",
       caption = "SR2 = pseudo R2; O. = counts truncated to occurrences; C. = statistic conditional upon occurrence. ") +
  theme_clean()

p_fit
ggsave(p_fit, filename = "cash.species.p_fit.pdf", 
       width = 8, height=11, bg="white")

mean(MF$SR2, na.rm=T)

sbquants <- summary(mpost$Beta)$quantiles %>%
  as_tibble(rownames = "variable") %>% 
  mutate(sign = `2.5%` * `97.5%`) %>%
  filter(sign>0) %>%
  tidyr::separate(variable,
           into = c("variable", "species"),
           sep = ",") %>%
  mutate(variable = str_sub(variable, 3,nchar(variable)-5),
         species = str_sub(species, 2,nchar(species)-6) %>% trimws) %>%
  filter(variable!= "(Intercept)") %>%
  dplyr::select(variable,species,`2.5%`,`50%`,`97.5%`) %>%
  arrange(variable)

vp_df <- VP$vals%>%
  as_tibble(rownames = "variable") %>%
  pivot_longer(cols=names(.)[2:ncol(.)], 
               names_to = "Species", 
               values_to = "value") %>%
  na.omit()

vp_summary <- vp_df %>%
  group_by(variable) %>%
  summarise(value_pct = mean(value) * 100,
            value_pct = round(value_pct, 1)) %>%
  ungroup() 

vp_order <- vp_df %>%
  filter(variable == "ClimateZ") %>%
  #arrange(value) %>%
  mutate(Species_f = factor(Species, levels = .$Species)) %>%
  dplyr::select(Species, Species_f) 

colors <- paletteer::paletteer_d("ggsci::default_igv", 30)
vp <- left_join(vp_df, vp_order) %>% 
  left_join(vp_summary) %>%
  mutate(variable = paste0(variable, " (",value_pct," %)")) %>%
  ggplot(aes(x=value, y=Species_f, fill = variable)) +
  geom_bar(stat="identity", color="black") +
  #theme_classic() +
  ylab("Species") +
  xlab("Proportion of Variance Explained") +
  scale_fill_manual(values=as.vector(colors), name="Variable (Avg % Explained)")
  #scale_fill_discrete() +
  #theme(legend.position = "right",
  #      legend.text = element_markdown(),
  #      legend.title = element_blank(),
  #      # legend.justification = c(1,0),
  #      legend.background = element_rect(color="black")) +
  #ggtitle("Variance Partitioning")

vp
ggsave(vp, filename = "cash.species.vp.pdf", 
       width = 8, height=11, bg="white")

```

#<eof>