---
title: "Machine Learning Graphing"
author: "Heather Deel"
date: "2023-09-18"
output: html_document
---

### Libraries
### make sure to clean these up, get rid of ones I don't use
```{r}
library(ggplot2)
library(readxl)
library(ggpubr)
library(paletteer)
library(tidyverse)
library(sjPlot)
library(cowplot)
library(janitor)
library(phyloseq)
library(M3C)
library(vegan)
library(cowplot)
library(grid)
library(gridExtra)
library(pairwiseAdonis)
library(ggcorrplot)
library(Hmisc)
library(rlist)
library(pipeR)
library(RColorBrewer)
library(png)
library(gplots)
library(dplyr)
library(ggrepel)
library(ggforce)
library(egg)
library(magick) 

# function to set the panel size of a multi-panel ggplot object
set_panel_size <- function (p=NULL, g=ggplotGrob(p), 
                            file=NULL, margin = unit(1,"mm"), 
                            width=unit(4, "cm"), height=unit(4, "cm")) {
  panels <- grep("panel", g$layout$name)
  panel_index_w<- unique(g$layout$l[panels])
  panel_index_h<- unique(g$layout$t[panels])
  nw <- length(panel_index_w)
  nh <- length(panel_index_h)
  
  if(getRversion() < "3.3.0") {
    g$widths <- grid:::unit.list(g$widths)
    g$heights <- grid:::unit.list(g$heights)
    g$widths[panel_index_w] <-  rep(list(width),  nw)
    g$heights[panel_index_h] <- rep(list(height), nh)
  } else {
    g$widths[panel_index_w] <-  rep(width,  nw)
    g$heights[panel_index_h] <- rep(height, nh)
  }
  if (!is.null(file)) {
    ggsave(file, g,
           width = convertWidth(sum(g$widths) + margin, unitTo = "in", valueOnly = TRUE),
           height = convertHeight(sum(g$heights) + margin, unitTo = "in", valueOnly = TRUE))
  }
  invisible(g)  
}
```

### Figure X
# comparison of model observed vs predicted R^2 values
```{r}
ml_stats <- read.csv("machine_learning/16S_EC/All_stats.csv")
ml_stats2 <- filter(ml_stats, model_type == "RA" | model_type == "Total" | model_type == "RA_TAX")

#stat.test <- compare_means(r2_val ~ model_type,
#                           data = ml_stats2,
#                           method = "wilcox.test",
#                           p.adj='fdr', group.by = "myVar")

#stat.test$Comparison <- paste0(stat.test$myVar, " - ", stat.test$group1)
#groups <- rcompanion::cldList(p.adj ~ Comparison,
#                              data = stat.test,
#                              threshold = 0.05)
#groups$y <- 1
#groups <- groups[1:13,]
#groups$p.signif <- stat.test$p.signif

p1 <- ggplot(ml_stats2, aes(x=myVar, y=r2_val, fill = model_type)) + 
  geom_boxplot() + 
  xlab("Indicator") +
  ylab(bquote(R^2~"Values")) +
  theme_bw() + 
  ylim(0,1) +
  theme(axis.text.x=element_text(angle=45, hjust=1)) +
  scale_fill_discrete(name = "Data Type", labels = c("RA EC", "RA TAX", "Total EC")) +
  scale_x_discrete(labels = c("ACE","Active C","Agg. Stab.","Fe","K","Mg","Mn","P",
                              "pH","Respiration","SOM","Water Capacity","Zn"))
  #geom_text(inherit.aes=F, data=groups, aes(x=Group, y=y, label=p.signif))
p1

#ggsave("figures/ml_r2.png", unit = "in", width = 10, height = 5, dpi = 300, device = "png")

# calculating sig differences between EC RA and TAX RA for manuscript purposes
ml_stats3 <- filter(ml_stats, model_type == "RA" | model_type == "RA_TAX")

stat.test <- compare_means(r2_val ~ model_type,
                           data = ml_stats3,
                           method = "wilcox.test",
                           p.adj='fdr', group.by = "myVar")

stat.test$Comparison <- paste0(stat.test$myVar, " - ", stat.test$group1)
groups <- rcompanion::cldList(p.adj ~ Comparison,
                              data = stat.test,
                              threshold = 0.05)
groups$y <- 1
groups <- groups[1:13,]
groups$p.signif <- stat.test$p.signif

# calculate means for manuscript purposes
means <- aggregate(r2_val ~  model_type+myVar, ml_stats3, mean)
means2 <- means %>% 
  group_by(myVar) %>% 
  mutate(r2_diff = )

p2 <- ggplot(ml_stats3, aes(x=myVar, y=r2_val, fill = model_type)) + 
  geom_boxplot() + 
  xlab("Indicator") +
  ylab(bquote(R^2~"Values")) +
  theme_bw() + 
  ylim(0,1) +
  theme(axis.text.x=element_text(angle=45, hjust=1)) +
  scale_fill_discrete(name = "Data Type", labels = c("RA EC", "RA TAX")) +
  #scale_x_discrete(labels = c("ACE","Active C","Agg. Stab.","Fe","K","Mg","Mn","P","pH","Respiration","SOM","Water Capacity","Zn")) +
  geom_text(inherit.aes=F, data=groups, aes(x=Group, y=y, label=p.signif))
p2

# calculating sig differences between EC RA and EC TOTAL for manuscript purposes
ml_stats4 <- filter(ml_stats, model_type == "RA" | model_type == "Total")

stat.test <- compare_means(r2_val ~ model_type,
                           data = ml_stats4,
                           method = "wilcox.test",
                           p.adj='fdr', group.by = "myVar")

stat.test$Comparison <- paste0(stat.test$myVar, " - ", stat.test$group1)
groups <- rcompanion::cldList(p.adj ~ Comparison,
                              data = stat.test,
                              threshold = 0.05)
groups$y <- 1
groups <- groups[1:13,]
groups$p.signif <- stat.test$p.signif

p3 <- ggplot(ml_stats4, aes(x=myVar, y=r2_val, fill = model_type)) + 
  geom_boxplot() + 
  xlab("Indicator") +
  ylab(bquote(R^2~"Values")) +
  theme_bw() + 
  ylim(0,1) +
  theme(axis.text.x=element_text(angle=45, hjust=1)) +
  #scale_fill_discrete(name = "Data Type", labels = c("RA EC", "Total EC")) +
  #scale_x_discrete(labels = c("ACE","Active C","Agg. Stab.","Fe","K","Mg","Mn","P",
  #                            "pH","Respiration","SOM","Water Capacity","Zn"))
  geom_text(inherit.aes=F, data=groups, aes(x=Group, y=y, label=p.signif))
p3
```

### Create list of taxa in all models sorted by mean VIP
```{r}
myVars <- c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'zn', 'ace', 'activeC', 'ph', 'p', 'k','fe')
for (myVar in myVars) {
  filenames <- list.files(paste0("machine_learning/16S_TAX/", myVar,"_model_results_RA_TAX/"), 
                          pattern = paste0(myVar,"_RA_TAX_varimp*"), full.names = TRUE)
  varimps <- lapply(filenames, read.csv, header = FALSE)
  varimps_full <- varimps %>%
    reduce(full_join)
  names(varimps_full) <- c('Predictor', 'VIP', 'Run')

 gDF <- varimps_full %>%
   filter(Predictor != "ClimateZ" & Predictor != "clay" & Predictor != "DNA") %>%
   group_by(Predictor) %>%
   summarise(VIP = signif(mean(VIP),3),
             n=n()) %>%
   filter(n == 25)
  gDF$indicator <- myVar
  
  write.table(gDF, "machine_learning/16S_TAX/top_TAX_All.csv", sep = ",",
              append = TRUE, row.names = FALSE, col.names = FALSE)
}
```

### Create list of enzymes in all models sorted by mean VIP
```{r}
myVars <- c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'zn', 'ace', 'activeC', 'ph', 'p', 'k','fe')
for (myVar in myVars) {
  filenames <- list.files(paste0("machine_learning/16S_EC/", myVar,"_model_results_RA_corr/"), 
                          pattern = paste0(myVar,"_RA_corr_varimp*"), full.names = TRUE)
  varimps <- lapply(filenames, read.csv, header = FALSE)
  varimps_full <- varimps %>%
    reduce(full_join)
  names(varimps_full) <- c('Predictor', 'VIP', 'Run')

 gDF <- varimps_full %>%
   filter(Predictor != "ClimateZ" & Predictor != "clay" & Predictor != "DNA") %>%
   group_by(Predictor) %>%
   summarise(VIP = signif(mean(VIP),3),
             n=n()) %>%
   filter(n == 25)
  
  gDF$indicator <- myVar
  
  write.table(gDF, "machine_learning/16S_EC/top_ECs_All.csv", sep = ",",
              append = TRUE, row.names = FALSE, col.names = FALSE)
}
```

### Get PICRUSt2 copy number data, parse taxonomy, create top_TAX and top_EC files for each indicator
```{r}
### import important ECs
top_ECs <- read.csv("machine_learning/16S_EC/top_ECs_All.csv", header = FALSE)
colnames(top_ECs) <- c("EC", "EC.VIP", "Runs", "Indicator")
top_ECs$EC <- gsub("EC:", "EC.", top_ECs$EC) 
top_ECs$EC <- gsub("_", "\\.", top_ECs$EC) 

### import important Taxa
top_TAX <- read.csv("machine_learning/16S_TAX/top_TAX_All.csv", header = FALSE)
colnames(top_TAX) <- c("species", "TAX.VIP", "Runs", "Indicator")

### import picrust2 copy number data for each data set - to add to top_TAX
hops2018.picrust <- read_tsv("picrust2_files/Hops.2018/EC_predicted_and_nsti.tsv")
hops2018.picrust$study <- "hops.2018"
hopsARS.picrust <- read_tsv("picrust2_files/Hops.ARS/EC_predicted_and_nsti.tsv")
hopsARS.picrust$study <- "hops.ARS"
NRCS.picrust <- read_tsv("picrust2_files/NRCS/EC_predicted_and_nsti.tsv")
NRCS.picrust$study <- "NRCS"
rangeland.picrust <- read_tsv("picrust2_files/Rangeland/EC_predicted_and_nsti.tsv")
rangeland.picrust$study <- "rangeland"

picrust.all <- bind_rows(hops2018.picrust, hopsARS.picrust, NRCS.picrust, rangeland.picrust)
picrust.all$sequence <- gsub("OTU", "", picrust.all$sequence)  # get rid of "OTU" in taxon column
colnames(picrust.all)[1] <- "OTU_ID"

# select only picrust data for ECs of interest
wanted <- c("study", "OTU_ID", unique(top_ECs$EC))
picrust.select <- picrust.all[,wanted]
picrust.select$OTU_ID <- as.numeric(picrust.select$OTU_ID)

# get taxonomy data from all runs
hops2018.otuid <- read.delim("16S_dada2_output/Hops.2018.taxonomy.otuid.csv", sep = ",")
hops2018.otuid$study <- "hops.2018"
hopsARS.otuid <- read.delim("16S_dada2_output/Hops.ARS.taxonomy.otuid.csv", sep = ",")
hopsARS.otuid$study <- "hops.ARS"
NRCS.otuid <- read.delim("16S_dada2_output/NRCS.taxonomy.otuid.csv", sep = ",")
NRCS.otuid$study <- "NRCS"
rangeland.otuid <- read.delim("16S_dada2_output/Rangeland.taxonomy.otuid.csv", sep = ",")
rangeland.otuid$study <- "rangeland"
otuid <- rbind(hops2018.otuid, hopsARS.otuid, NRCS.otuid, rangeland.otuid)

otuid <- otuid %>%
  separate(species, into=c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep=";") %>%
  select(study, tax_ID, OTU_ID, Kingdom, Phylum, Class, Order, Family, Genus, Species) %>%
  mutate(Species2 = Species) %>%
  mutate(Species = gsub("\\(", "_", Species)) %>%
  mutate(Species = gsub("\\)", "_", Species)) %>%
  mutate(Species = gsub("\\.", "_", Species)) %>%
  mutate(Species = gsub("\\-", "_", Species)) %>%
  mutate(Species = gsub(" ", "_", Species))

top_TAX <- top_TAX %>%
  left_join(otuid, by=c("species" = "Species"))

top_TAX <- top_TAX %>%
  left_join(picrust.select, by=c("study", "OTU_ID")) %>%
  drop_na() %>%
  select(-study, -OTU_ID, -tax_ID, -species)

top_TAX <- top_TAX %>%
  group_by(Indicator, Kingdom, Phylum, Class, Order, Family, Genus, Species2) %>%
  summarise_if(is.numeric, mean, na.rm=TRUE) %>%
  ungroup()

saveRDS(top_ECs, 'top_ECs.RDS')
saveRDS(top_TAX, 'top_TAX.RDS')
#_________________________

myVars <- c("ace", "activeC", "agg_stab", "fe", "k", "mg", "mn", "p", "ph", "resp", "SOM", "water_cap", "zn")
for (myVar in myVars) {
  # subset data by indicator, taxa, and ECs of interest
  top_ECs.sub <- top_ECs %>%
    filter(Indicator == myVar) %>%
    arrange(desc(EC.VIP))
  write.csv(top_ECs.sub, paste0("figures/top_predictors/top_ECs.", myVar, ".csv"))
  
  top_TAX.sub <- top_TAX %>%
    filter(Indicator == myVar) %>%
    arrange(desc(TAX.VIP)) %>%
    select(Indicator, Kingdom, Phylum, Class, Order, Family, Genus, Species2, 
           TAX.VIP, Runs, any_of(top_ECs.sub$EC))
  write.csv(top_TAX.sub, paste0("figures/top_predictors/top_TAX.", myVar, ".csv"))
}
```

### Figure X
# functional redundancy in taxonomic models
```{r}
top_ECs <- readRDS('scripts/top_ECs.RDS')
top_TAX <- readRDS('scripts/top_TAX.RDS')

### graphs
myVars <- c("ace", "activeC", "agg_stab", "fe", "k", "mg", "mn", "p", "ph", "resp", "SOM", "water_cap", "zn")
pl <- list()
for (myVar in myVars) {
  # changing myVar format for cleaner plot titles
  if (myVar == 'resp'){
    myTitle <- "Resp"
  }
  if (myVar == 'agg_stab'){
    myTitle <- "AggStab"
  }
  if (myVar == 'water_cap'){
    myTitle <- "WaterCap"
  }
  if (myVar == 'mg'){
    myTitle <- "Mg"
  }
  if (myVar == 'mn'){
    myTitle <- "Mn"
  }
  if (myVar == 'zn'){
    myTitle <- "Zn"
  }
  if (myVar == 'ace'){
    myTitle <- "ACE"
  }
  if (myVar == 'activeC'){
    myTitle <- "ActiveC"
  }
  if (myVar == 'ph'){
    myTitle <- "pH"
  }
  if (myVar == 'p'){
    myTitle <- "P"
  }
  if (myVar == 'k'){
    myTitle <- "K"
  }
  if (myVar == 'fe'){
    myTitle <- "Fe"
  }
  if (myVar == 'SOM'){
    myTitle <- "SOM"
  }
  
  # subset data by indicator, taxa, and ECs of interest
  indicator.ECs <- top_ECs %>%
    filter(Indicator == myVar) %>%
    arrange(desc(EC.VIP)) %>%
    top_n(10, EC.VIP)

  indicator.TAX <- top_TAX %>%
    filter(Indicator == myVar) %>%
    arrange(desc(TAX.VIP)) %>%
    select(Indicator, Kingdom, Phylum, Class, Order, Family, Genus, Species2, 
           TAX.VIP, Runs, any_of(indicator.ECs$EC)) %>%
    top_n(10, TAX.VIP)
  
  indicator.TAX.long <- indicator.TAX %>%
    pivot_longer(cols=starts_with("EC."), names_to="EC", values_to="Copies")
  
  gDF <- indicator.ECs %>%
    left_join(indicator.TAX.long, by="EC")
  gDF["Copies"][gDF["Copies"] == 0] <- NA
  
  # create heatmaps
  library(forcats)
  p2 <- ggplot(gDF, aes(x = fct_reorder(EC, EC.VIP, .desc=TRUE, .na_rm=FALSE), 
                        y = fct_reorder(Species2, TAX.VIP, .desc=TRUE, .na_rm=FALSE), fill = Copies)) +
    geom_tile(color='grey20') +
    #geom_text(aes(label=round(Copies,3)), size=2) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, size = 10),
          axis.text.y = element_text(size = 10),
          legend.text = element_text(size = 10)) +
    scale_fill_binned(type="viridis", direction=-1,
                      breaks = c(1,2,3,4,5,6),
                      #labels=c("(0,1]","(1,2]","(2,3]","(3,4]",
                      #         "(4,5]","(5,6]"),
                      limits=c(0,6)) +
  labs(x = "", y = "", title = myTitle, fill = "Enzyme Copies")
  
  # generating individual plots so I can arrange them with the legend
  assign(paste0(myVar,"_hm"),p2)
  
  # if (myTitle == "SOM") {
  #   legend = get_legend(p2)
  #   ggsave(plot=legend, "figures/heatmap_dkm/legend.png", device='png', dpi=600,
  #          height=3, width=2)
  #   legend2 <- as_ggplot(legend)
  # }
  
  #p2 <- p2 + theme(legend.position = "none")
  #p2 <- set_panel_size(p2, height=unit(3, "in"), width=unit(3, "in"))
  #ggsave(paste0("figures/heatmap_dkm/ez_copies_per_species_heatmap_", myTitle,".png"), 
  #       device='png', dpi=600, height=4, width=6)
  #pl[[myTitle]] <- p2
}

# create merged plot of heatmaps
#p_heatmap_all <- marrangeGrob(pl, legend2, top=NULL, layout_matrix = matrix(1:16,  nrow = 5, ncol=3, byrow=TRUE))

#ggsave(plot=p_heatmap_all, "figures/heatmap_dkm/ez_copies_per_species_heatmap.png", 
#       device='png', dpi=600, height=22, width=17.5)

p_heatmap_leg <- ggpubr::ggarrange(ace_hm, activeC_hm, agg_stab_hm, fe_hm, k_hm, mg_hm,
                           mn_hm, p_hm, ph_hm, resp_hm, SOM_hm, water_cap_hm, zn_hm,
                           nrow = 5, ncol = 3, common.legend = TRUE)
p_heatmap_leg

p_heatmap_leg <- annotate_figure(p_heatmap_leg, left = text_grob("Important Taxa", rot = 90, vjust = 1, size = 14),
                bottom= text_grob("Important Enzymes", size = 14))

ggsave(plot=p_heatmap_leg, "figures/heatmap_dkm/ez_copies_per_species_heatmap2.png", 
       device='png', dpi=600, height=22, width=17.5)
```

### Figure X
# PCOAs using tsne
```{r}
### EC data PCOA
P <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')

data <- P[,3:2438]

meta <- P
meta <- meta[,c(1:2,2439:2587)]

p1 <- M3C::tsne(t(data), labels=FALSE, dotsize = 0, seed=1234, perplex=15) + theme_bw() +
  geom_point(aes(fill=meta$ClimateZ, shape=meta$Current_Land_Use), size=4) +
  scale_shape_manual(values=c(21,22,23,24), labels = c("Annual Cropland","Perennial Cropland","Rangeland")) +
  guides(size="none",
         fill = guide_legend(override.aes = list(size = 4, shape=21))) +
  labs(fill="Climate Zone", shape="Land Use", x="tSNE 1", y="tSNE 2", title = "Enzyme Data")
p1
#ggsave(paste0("figures/All_enzymes_RA_corr_tSNE.png"), device='png', dpi=600, height=4, width=6)

# permanova between land use and climate zone
permanova_land <- pairwise.adonis2(data ~ Current_Land_Use,
                                data = meta,
                                permutations = 9999)
# p-vals for all comparisons less than 0.05
mapply(write.table, x = permanova_land, file = paste("figures/climate_land_permanova/ez_",names(permanova_land), "txt", sep = "."), MoreArgs = list(row.names=FALSE, sep = "\t"))

permanova_climate <- pairwise.adonis2(data ~ ClimateZ,
                                data = meta,
                                permutations = 9999)
# p-vals for all comparisons less than 0.05
mapply(write.table, x = permanova_climate, file = paste("figures/climate_land_permanova/ez_",names(permanova_climate), "txt", sep = "."), MoreArgs = list(row.names=FALSE, sep = "\t"))

### taxa data PCOA
P <- readRDS('machine_learning/16S_TAX/ml_TAX_16S.RDS')

data <- P[,152:7614]

meta <- P
meta <- meta[,1:151]

p2 <- M3C::tsne(t(data), labels=FALSE, dotsize = 0, seed=1234, perplex=15) + theme_bw() +
  geom_point(aes(fill=meta$ClimateZ, shape=meta$Current_Land_Use), size=4) +
  scale_shape_manual(values=c(21,22,23,24), labels = c("Annual Cropland","Perennial Cropland","Rangeland")) +
  guides(size="none",
         fill = guide_legend(override.aes = list(size = 4, shape=21))) +
  labs(fill="Climate Zone", shape="Land Use", x="tSNE 1", y="tSNE 2", title = "Taxa Data")
p2
#ggsave(paste0("figures/TAX_RA_tSNE.png"), device='png', dpi=600, height=4, width=6)

# permanova between land use and climate zone
permanova_land <- pairwise.adonis2(data ~ Current_Land_Use,
                                data = meta,
                                permutations = 9999)
# p-vals for all comparisons less than 0.05
mapply(write.table, x = permanova_land, file = paste("figures/climate_land_permanova/tax_",names(permanova_land), "txt", sep = "."),MoreArgs = list(row.names=FALSE, sep = "\t"))

permanova_climate <- pairwise.adonis2(data ~ ClimateZ,
                                data = meta,
                                permutations = 9999)
# p-vals for all comparisons less than 0.05
mapply(write.table, x = permanova_climate, file = paste("figures/climate_land_permanova/tax_",names(permanova_climate), "txt", sep = "."),MoreArgs = list(row.names=FALSE, sep = "\t"))

# can output permanova results to a file to compare how well enzyme data and taxa data distinguishes between land use and climate zones - compare R^2 and F statistics

p_pcoa <- ggarrange(p1, p2, common.legend = TRUE, nrow = 1, ncol = 2)
p_pcoa
#ggsave(paste0("figures/EC_TAX_RA_tSNE.png"), device='png', dpi=600, height=4, width=8.2)
```

### Figure X and Figure X
# Feature importance distributions of indicators with all enzyme models
# separating by positively (Figure 4) and negatively (Figure 5) correlated enzymes
```{r}
top_ECs <- readRDS('scripts/top_ECs.RDS')

# merge top_ECs with GIBBs list to add context for talking about in the manuscript
gibbs <- read.csv("picrust2_files/GIBBs.EC.numbers.csv")
gibbs$EC <- gsub("EC:","EC.",gibbs$EC)
top_ECs_gibbs <- top_ECs %>% 
  left_join(gibbs, by = "EC") %>% 
  filter(!is.na(Category))

myVars <- c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'zn', 'ace', 'activeC', 'ph', 'p', 'k','fe')

for (myVar in myVars) {
 indicator.ECs <- top_ECs %>%
    filter(Indicator == myVar) %>%
    arrange(desc(EC.VIP))

 # calculate positive and negative correlations between enzymes and indicators
 ml_EC_RA_corr <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')
 colnames(ml_EC_RA_corr) <- gsub("EC:", "EC.", colnames(ml_EC_RA_corr))
 
 ml_indic <- ml_EC_RA_corr %>% 
   select(indicator.ECs$EC, paste0(myVar))
 
 ml_indic <- as.matrix(ml_indic)
 ec.mat <- rcorr(ml_indic)
 ec.r <- ec.mat$r
 ec.indic <- ec.r %>% 
   subset(select = paste0(myVar))
 ec.indic <- data.frame(ec.indic)

 # format ec.indic but separate by positive and negative correlation values
 ec.indic.pos <- ec.indic %>% 
   filter(ec.indic[,1] > 0)
 ec.indic.neg <- ec.indic %>% 
   filter(ec.indic[,1] < 0)
 ec.indic.pos <- rownames_to_column(ec.indic.pos, var = "EC")
 ec.indic.neg <- rownames_to_column(ec.indic.neg, var = "EC")
 ec.indic.pos <- ec.indic.pos[ec.indic.pos$EC != myVar,]
 ec.indic.neg <- ec.indic.neg[ec.indic.neg$EC != myVar,]

  # filter out poor indicators in pos and neg correlations
 gDF.pos <- indicator.ECs %>% 
   filter(EC %in% ec.indic.pos$EC) 
 gDF.neg <- indicator.ECs %>% 
   filter(EC %in% ec.indic.neg$EC) 
 
  temp.pos <- gDF.pos %>% group_by(EC) %>%
    summarise(Var = median(EC.VIP)) %>%
    top_n(10, Var) 
  temp.neg <- gDF.neg %>% group_by(EC) %>%
    summarise(Var = median(EC.VIP)) %>%
    top_n(10, Var) 
  
  gDF.pos <- gDF.pos %>%
    filter(EC %in% temp.pos$EC)
  gDF.neg <- gDF.neg %>%
    filter(EC %in% temp.neg$EC)
  
  # add in gibbs info for context. Assign to new variable for each indicator
  gDF.pos <- gDF.pos %>% 
    left_join(gibbs, by = "EC")
  assign(paste0(myVar,"_gibbs_pos"),gDF.pos)
  gDF.neg <- gDF.neg %>% 
    left_join(gibbs, by = "EC")
  assign(paste0(myVar,"_gibbs_neg"),gDF.neg)
  
  # changing myVar format for cleaner plot titles
  if (myVar == 'resp'){
    myVar <- "Resp"
  }
  if (myVar == 'agg_stab'){
    myVar <- "AggStab"
  }
  if (myVar == 'water_cap'){
    myVar <- "WaterCap"
  }
  if (myVar == 'mg'){
    myVar <- "Mg"
  }
  if (myVar == 'mn'){
    myVar <- "Mn"
  }
  if (myVar == 'zn'){
    myVar <- "Zn"
  }
  if (myVar == 'ace'){
    myVar <- "ACE"
  }
  if (myVar == 'activeC'){
    myVar <- "ActiveC"
  }
  if (myVar == 'ph'){
    myVar <- "pH"
  }
  if (myVar == 'p'){
    myVar <- "P"
  }
  if (myVar == 'k'){
    myVar <- "K"
  }
  if (myVar == 'fe'){
    myVar <- "Fe"
  }
  # positive correlation graph
  means <- aggregate(EC.VIP ~ EC, gDF.pos, median)
  means$EC.VIP <- round(means$EC.VIP, 3)
  p.pos <- ggplot(gDF.pos, aes(fct_reorder(EC, EC.VIP, .desc=TRUE, .na_rm=FALSE), y=EC.VIP)) +
    geom_bar(stat = "identity", width = 0.75) + 
    theme_bw() +
    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5), 
          legend.position = "none", plot.title = element_text(size = 12),
          plot.margin = unit(c(0.05,0.05,0.05,0.05), "cm")) +
    labs(x="", y="", title = myVar)
  p.pos
  assign(paste0(myVar,"_pos"),p.pos)
  #ggsave(paste0("figures/varimps/", myVar, "_RA_corr_varimp.pos.png"), device='png', dpi=600, height=4, width=6)
  
  # negative correlation graph
  means <- aggregate(EC.VIP ~ EC, gDF.neg, median)
  means$EC.VIP <- round(means$EC.VIP, 3)
  p.neg <- ggplot(gDF.neg, aes(fct_reorder(EC, EC.VIP, .desc=TRUE, .na_rm=FALSE), y=EC.VIP)) +
    geom_bar(stat = "identity", width = 0.75) + 
    theme_bw() +
    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5), 
          legend.position = "none", plot.title = element_text(size = 12),
          plot.margin = unit(c(0.05,0.05,0.05,0.05), "cm")) +
    labs(x="", y="", title = myVar)
  p.neg
  assign(paste0(myVar,"_neg"),p.neg)
  #ggsave(paste0("figures/varimps/", myVar, "_RA_corr_varimp.neg.png"), device='png', dpi=600, height=8, width=8)
}

# positive correlations graph
p_merged_pos <- ggarrange(ACE_pos, ActiveC_pos, AggStab_pos, 
                          Fe_pos, K_pos, Mg_pos, Mn_pos, 
                          P_pos, pH_pos, Resp_pos, 
                          SOM_pos, WaterCap_pos, Zn_pos, 
                          nrow = 2, ncol = 7)
p_merged_pos

annotate_figure(p_merged_pos, left = text_grob("Average Variable Importance", rot = 90, vjust = 1, size = 14))

#ggsave("figures/all_enzyme_varimps_pos.png", device='png', dpi=600, height=6, width=12)

# negative correlations graph
p_merged_neg <- ggarrange(ACE_neg, ActiveC_neg, AggStab_neg, 
                          Fe_neg, K_neg, Mg_neg, Mn_neg, 
                          P_neg, pH_neg, Resp_neg, 
                          SOM_neg, WaterCap_neg, Zn_neg,
                          nrow = 2, ncol = 7)
p_merged_neg

annotate_figure(p_merged_neg, left = text_grob("Average Variable Importance", rot = 90, vjust = 1, size = 14))

#ggsave("figures/all_enzyme_varimps_neg.png", device='png', dpi=600, height=6, width=12)
```

### Table SX
# summary statistics of number of features included in each model
```{r}
myVars <- c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'zn', 'ace', 'activeC', 'ph', 'p', 'k','fe')

for (myVar in myVars) {
  filenames <- list.files(paste0("machine_learning/16S_EC/", myVar,"_model_results_RA_corr/"), 
                          pattern = paste0(myVar,"_RA_corr_varimp*"), full.names = TRUE)
  varimps <- lapply(filenames, read.csv, header = FALSE)
  varimps_full <- varimps %>%
    reduce(full_join)
  names(varimps_full) <- c('Predictor', 'VIP', 'Run')
  varimps_full <- varimps_full %>% 
    filter(Predictor != "ClimateZ" & Predictor != "clay" & Predictor != "DNA")
  varimps_full_sum <- varimps_full %>% 
    group_by(Run) %>% 
    summarise(feature_count = n())
  var_range <- range(varimps_full_sum$feature_count)
  var_range <- data.frame(var_range)
  var_range <- t(var_range)
  var_range$indicator <- myVar
  var_range <- data.frame(var_range)
  colnames(var_range) <- c("min_vars", "max_vars","indicator")
  write.table(var_range, "machine_learning/16S_EC/ECs_per_model_range.csv", sep = ",",
              append = TRUE, row.names = FALSE, col.names = FALSE)
}
```

### Figure X
# Unrooted dendrogram
# PCA with biplot showing which metabolisms drive indicators
####### Decide which of the figures in this section make sense to go in the manuscript
####### I like the idea of one biplot that for each indicator, it shows which level 2 metabolism drives it the most - can generate and put in supp materials
####### The biplot can perhaps be accompanied by a dendrogram?

```{r}
### dendrogram
df <- read_xlsx("figures/top_predictors/Summary.xlsx", sheet="All_NZ", range="A2:O57")

data <- df %>%
  select(-Level2, -Level3) %>%
  as.data.frame()

data <- t(data)
colnames(data) <- df$Level3
data <- data[,colSums(data) > 0]
#data <- decostand(data, method="pa")

meta <- data.frame(Level2=df$Level2, Level3=df$Level3)
meta <- meta[meta$Level3 %in% colnames(data),]

dd <- dist(data, method = "euclidean")
hc <- hclust(dd, method = "ward.D2")
hcd <- as.dendrogram(hc)

plot(ape::as.phylo(hc), type = "unrooted", cex = 0.6, no.margin = TRUE)
# manually saved as png to "C:\Users\Heather.Deel\OneDrive - USDA\microbial_indicators\microbial_indicators\microbial_indicators\figures\metabolisms_SHMI\unrooted_dendrogram.png"

### ordination of all data with all indicators, driven by pathways
ord <- metaMDS(data)

q <- ordiplot(ord, choices=c(1,2), scaling=1, type='none') 
species <- data.frame(q$species)
names(species) <- c('Axis1', 'Axis2')
species$Level2 <- meta$Level2
wanted <- goeveg::ordiselect(data, ord, fitlim=0.4)
species <- species[wanted,]
species$label <- row.names(species)
species$label <- gsub("\\[", "(", species$label)
species$label <- gsub("\\]", ")", species$label)
species$label <-  stringr::str_extract(string = species$label, pattern = "(?<=\\().*(?=\\))")
species$label <- gsub("TCA cycle\\) \\(", "", species$label)
sites <- data.frame(q$sites)
names(sites) <- c('Axis1', 'Axis2')

p2 <- ggplot(species) +
  geom_point(data=sites, inherit.aes=F, aes(x=Axis1, y=Axis2)) +
  ggrepel::geom_text_repel(data=sites, inherit.aes=F, 
                  aes(x=Axis1, y=Axis2, label=row.names(sites))) +
  geom_segment(aes(x=0, y=0, xend=Axis1, yend=Axis2, color=Level2), 
               arrow = arrow(length = unit(0.1, "cm"))) +
  ggrepel::geom_text_repel(aes(x=Axis1, y=Axis2, label = species$label, color=Level2), size=3, segment.color="grey80", max.overlaps = 20) +
  #ggforce::geom_circle(aes(x0=0, y0=0, r=1)) +
  coord_equal() +
  theme_bw() +
  labs(color = "KO Pathway")
p2

# read in dendrogram, merge with p2, and save
dendro <- image_read("figures/metabolisms_SHMI/unrooted_dendrogram.png") %>% 
  image_ggplot()

dendro_pathways <- ggarrange(dendro, p2, ncol = 2, nrow = 1, labels = c("A","B"))
dendro_pathways

# ggsave for some reason will only save the dendrogram of dendro_pathways
# manually save as png at figures/dendro_pathways.png
```

```{r}
### pcoa of land use, SHMI, and climate combined into one
### I think I should make one pcoa of just SHMI
### fig 3 in the manuscript is also already a pcoa with climate and land use
df <- read_xlsx("figures/top_predictors/Summary.xlsx", sheet="All_NZ", range="R2:AE805")
df$Enzyme <- gsub("EC.", "EC:", df$Enzyme)

P <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')

data <- P[,df$Enzyme]

meta <- P
meta <- meta[,c(1:2,2439:2587)]

ord <- metaMDS(data)

q <- ordiplot(ord, choices=c(1,2), scaling=3, type='none') 
species <- data.frame(q$species)
names(species) <- c('Axis1', 'Axis2')
wanted <- goeveg::ordiselect(data, ord, fitlim=0.1)
species <- species[wanted,]
species$label <- row.names(species)
sites <- data.frame(q$sites)
names(sites) <- c('Axis1', 'Axis2')

p4 <- ggplot(sites, aes(x=Axis1, y=Axis2)) +
  geom_point(aes(shape=meta$Current_Land_Use, fill=meta$ClimateZ, size=meta$SHMI2_bin)) +
  #facet_wrap(~ meta$ClimateZ) +
  scale_shape_manual(values=c(21,22,23,24), labels=c("Annual Cropland","Perennial Cropland","Rangeland")) +
  guides(fill=guide_legend(override.aes = list(size=4, shape=21))) +  #geom_text_repel(aes(label=row.names(sites))) +
  coord_equal() +
  theme_bw()
p4
```

### Figure X
# Grouping enzyme abundances by metabaolism and correlating with SHMI
####### may need to get rid of metabolic groups that are human related
```{r}
pathways <- read_xlsx("pathways_files/pathways.xlsx", sheet="pathways")
pathways <- pathways[,1:4]

df <- read_xlsx("figures/top_predictors/Summary.xlsx", sheet="All_NZ", range="R2:AE805")
df$Enzyme <- gsub("EC.", "EC:", df$Enzyme)

P <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')

data <- P[,df$Enzyme]
data_t <- t(data) %>% 
  data.frame()
data_t <- rownames_to_column(data_t, var = "EC")
data_t$EC <- gsub("EC.", "", data_t$EC)

df$Enzyme <- gsub("EC:", "", df$Enzyme)
pathways_filt <- pathways %>% 
  filter(EC %in% df$Enzyme)

data_t_pw <- data_t %>% 
  left_join(pathways, by = "EC") %>% 
  filter(!is.na(Level2)) %>% 
  group_by(Level2) %>% 
  summarise_if(is.numeric, sum, na.rm=TRUE)

data_pw <- t(data_t_pw)
data_pw <- row_to_names(data_pw, 1) %>% 
  data.frame()

data_pw2 <- bind_cols(data_pw, P)
data_pw2 <- data_pw2[,c(1:34,2472:2620)] # all data minus enzymes
data_pw3 <- data_pw2[,c(1:33,183)] # just level 2 metabolisms and SHMI2_bin

data_pw3 <- data_pw3 %>% 
  mutate_at(c(1:33), as.numeric)
colnames(data_pw3) <- gsub("\\.","_",colnames(data_pw3))

# change some colnames so they aren't so long and graph better
# tried various ways of wrapping the titles and none would work bc title is coming from m
colnames(data_pw3) <- gsub("Biosynthesis_of_other_secondary_metabolites","Biosynthesis_secondary_metabolites",colnames(data_pw3))
colnames(data_pw3) <- gsub("Cellular_community___eukaryotes","Cellular_community_eukaryotes",colnames(data_pw3))
colnames(data_pw3) <- gsub("Cellular_community___prokaryotes","Cellular_community_prokaryotes",colnames(data_pw3))
colnames(data_pw3) <- gsub("Folding__sorting_and_degradation","Folding_sorting_degradation",colnames(data_pw3))
colnames(data_pw3) <- gsub("Glycan_biosynthesis_and_metabolism","Glyc_biosynthesis_metabolism",colnames(data_pw3))
colnames(data_pw3) <- gsub("Information_processing_in_viruses","Info_processing_viruses",colnames(data_pw3))
colnames(data_pw3) <- gsub("Metabolism_of_cofactors_and_vitamins","Metabolism_cofactors_vitamins",colnames(data_pw3))
colnames(data_pw3) <- gsub("Metabolism_of_other_amino_acids","Metabolism_other_amino_acids",colnames(data_pw3))
colnames(data_pw3) <- gsub("Metabolism_of_terpenoids_and_polyketides","Metabolism_terpenoids_polyketides",colnames(data_pw3))
colnames(data_pw3) <- gsub("Signaling_molecules_and_interaction","Signaling_molecules_interaction",colnames(data_pw3))
colnames(data_pw3) <- gsub("Xenobiotics_biodegradation_and_metabolism","Xeno_biodegradation_metabolism",colnames(data_pw3))

saveRDS(data_pw3, "picrust2_files/level2_pw_SHMI.RDS")

metabolisms <- names(data_pw3)[1:33]

for (m in metabolisms) {
  p <- ggplot(data_pw3, aes_string(x = data_pw3$SHMI2_bin, y = m)) +
  geom_boxplot() +
  scale_x_discrete(labels = c("VL","L","M","H","VH")) +
  theme_bw() +
  theme(plot.title = element_text(size = 7, face = "plain"), plot.margin = unit(c(0.10,0.10,0.10,0.10), "cm")) +
  labs(x = "", y = "", title = m)
  #ggsave(paste0("figures/metabolisms_SHMI/SHMI_",m,".png"), device='png', dpi=600, height=4, width=6)
  assign(m,p)
}

# plot of all metabolic categories
plot_list <- list(Amino_acid_metabolism, Biosynthesis_secondary_metabolites,
                  Carbohydrate_metabolism, Cell_growth_and_death, Cell_motility,
                  Cellular_community_eukaryotes, Cellular_community_prokaryotes,
                  Chromosome, Circulatory_system, Development_and_regeneration,
                  Digestive_system, Endocrine_system, Energy_metabolism, Excretory_system, 
                  Folding_sorting_degradation, Glyc_biosynthesis_metabolism,
                  Immune_system, Info_processing_viruses, Lipid_metabolism,
                  Membrane_transport, Metabolism_cofactors_vitamins,
                  Metabolism_other_amino_acids, Metabolism_terpenoids_polyketides,
                  Nervous_system, Nucleotide_metabolism, Replication_and_repair,
                  Sensory_system, Signal_transduction, Signaling_molecules_interaction,
                  Transcription, Translation, Transport_and_catabolism,
                  Xeno_biodegradation_metabolism)
p_all <- ggarrange(plotlist = plot_list, nrow = 6, ncol = 6)
p_all

annotate_figure(p_all, left = text_grob("Relative Abundance", rot = 90, vjust = 1, size = 14),
                bottom = text_grob("SHMI Bin", size = 14))
 
ggsave("figures/metabolisms_SHMI/SHMI_all.png", device='png', dpi=600, height=10, width=13)
```

### Figure X
# subset to metabolisms that are significantly correlated with SHMI
```{r}
data_pw3 <- readRDS("picrust2_files/level2_pw_SHMI.RDS")
data_pw3$SHMI2_bin <- gsub("very low", "very_low", data_pw3$SHMI2_bin)
data_pw3$SHMI2_bin <- gsub("very high", "very_high", data_pw3$SHMI2_bin)

data_pw3$SHMI2_bin <- factor(data_pw3$SHMI2_bin, levels = c('very_low', 'low','medium', 'high', 'very_high'))

metabolisms <- names(data_pw3)[1:33]

for (m in metabolisms) {

formula <- paste(m,"~ SHMI2_bin")  
  
stat.test <- compare_means(as.formula(formula),
                           data = data_pw3,
                           method = "wilcox.test",
                           p.adj='fdr')

stat.test$Comparison <- paste0(stat.test$group1, " - ", stat.test$group2)
groups <- rcompanion::cldList(p.adj ~ Comparison,
                              data = stat.test,
                              threshold = 0.05)

 # create list of metabolisms with low abundance ranges so y can be assigned appropriately
 low_met <- c("Chromosome","Circulatory_system","Digestive_system","Excretory_system", 
              "Membrane_transport","Sensory_system")
 if (m %in% low_met) {
    groups$y <- max(data_pw3[,m]) + 0.25
 } else {
   groups$y <- max(data_pw3[,m]) + 5
 }

  p <- ggplot(data_pw3, aes_string(x = data_pw3$SHMI2_bin, y = m)) +
  geom_boxplot() +
  scale_x_discrete(labels = c("VL","L","M","H","VH")) +
  theme_bw() +
  theme(plot.title = element_text(size = 7, face = "plain"), plot.margin = unit(c(0.10,0.10,0.10,0.10), "cm")) +
  labs(x = "", y = "", title = m) +
  geom_text(inherit.aes=F, data=groups, aes(x=Group, y=y, label=Letter))
  p
  #ggsave(paste0("figures/metabolisms_SHMI/SHMI_",m,"stat.png"), device='png', dpi=600, height=4, width=6)
  assign(m,p)
}

# plot of all metabolic categories with significances
plot_list <- list(Amino_acid_metabolism, Biosynthesis_secondary_metabolites,
                  Carbohydrate_metabolism, Cell_growth_and_death, Cell_motility,
                  Cellular_community_eukaryotes, Cellular_community_prokaryotes,
                  Chromosome, Circulatory_system, Development_and_regeneration,
                  Digestive_system, Endocrine_system, Energy_metabolism, Excretory_system, 
                  Folding_sorting_degradation, Glyc_biosynthesis_metabolism,
                  Immune_system, Info_processing_viruses, Lipid_metabolism,
                  Membrane_transport, Metabolism_cofactors_vitamins,
                  Metabolism_other_amino_acids, Metabolism_terpenoids_polyketides,
                  Nervous_system, Nucleotide_metabolism, Replication_and_repair,
                  Sensory_system, Signal_transduction, Signaling_molecules_interaction,
                  Transcription, Translation, Transport_and_catabolism,
                  Xeno_biodegradation_metabolism)
p_all <- ggarrange(plotlist = plot_list, nrow = 6, ncol = 6)
p_all

annotate_figure(p_all, left = text_grob("Relative Abundance", rot = 90, vjust = 1, size = 14),
                bottom = text_grob("SHMI Bin", size = 14))
 
ggsave("figures/metabolisms_SHMI/SHMI_all_sig.png", device='png', dpi=600, height=10, width=13)

# subset to significant metabolic categories
plot_list2 <- list(Amino_acid_metabolism, Biosynthesis_secondary_metabolites,
                  Carbohydrate_metabolism, Cellular_community_prokaryotes,
                  Chromosome,Digestive_system, Endocrine_system, Energy_metabolism, 
                  Excretory_system, Folding_sorting_degradation, 
                  Glyc_biosynthesis_metabolism, Info_processing_viruses, Lipid_metabolism,
                  Membrane_transport, Metabolism_cofactors_vitamins,
                  Metabolism_other_amino_acids, Metabolism_terpenoids_polyketides,
                  Nucleotide_metabolism, Replication_and_repair,Transcription, 
                  Transport_and_catabolism,Xeno_biodegradation_metabolism)

p_sig <- ggarrange(plotlist = plot_list2, nrow = 6, ncol = 4)
p_sig

annotate_figure(p_sig, left = text_grob("Relative Abundance", rot = 90, vjust = 1, size = 14),
                bottom = text_grob("SHMI Bin", size = 14))
 
ggsave("figures/metabolisms_SHMI/SHMI_sig.png", device='png', dpi=600, height=11, width=10)
```






