---
title: "Machine Learning Graphing"
author: "Heather Deel"
date: "2023-09-18"
output: html_document
---

### Libraries
### make sure to clean these up, get rid of ones I don't use
```{r}
library(ggplot2)
library(readxl)
library(ggpubr)
library(paletteer)
library(tidyverse)
library(sjPlot)
library(cowplot)
library(janitor)
library(phyloseq)
library(M3C)
library(vegan)
library(cowplot)
library(grid)
library(gridExtra)
library(pairwiseAdonis)
library(ggcorrplot)
library(Hmisc)
library(rlist)
library(pipeR)
```

### Figure 1
# comparison of model observed vs predicted R^2 values
# need to rerun taxonomic models, with nutrient data
# then add to below plot
```{r}
# tax data was pulled from dan_out/All_stats_randomForest and added to All_stats.csv
ml_stats <- read.csv("machine_learning/16S_EC/All_stats.csv")
ml_stats_RA <- filter(ml_stats, RA_total_GIBBs == "RA" | RA_total_GIBBs == "RA_GIBBs")

stat.test <- compare_means(r2_val ~ RA_total_GIBBs,
                           data = ml_stats_RA,
                           method = "wilcox.test",
                           p.adj='fdr', group.by = "myVar")

stat.test$Comparison <- paste0(stat.test$myVar, " - ", stat.test$group1)
groups <- rcompanion::cldList(p.adj ~ Comparison,
                              data = stat.test,
                              threshold = 0.05)
groups$y <- 1
groups <- groups[1:13,]
groups$p.signif <- stat.test$p.signif

p1 <- ggplot(ml_stats_RA, aes(x=myVar, y=r2_val, fill = RA_total_GIBBs)) + 
  geom_boxplot() + 
  xlab("Indicator") +
  ylab(bquote(R^2~"Values")) +
  theme_bw() + 
  ylim(0,1) +
  theme(axis.text.x=element_text(angle=45, hjust=1)) +
  scale_fill_discrete(name = "Data Type", labels = c("RA", "RA GIBBs")) +
  scale_x_discrete(labels = c("ACE","Active C","Agg. Stab.","Fe","K","Mg","Mn","P",
                              "pH","Respiration","SOM","Water Capacity","Zn")) +
  geom_text(inherit.aes=F, data=groups, aes(x=Group, y=y, label=p.signif))
p1

ggsave("figures/ml_ec_r2_RA.png", unit = "in", width = 10, height = 5, dpi = 300, device = "pdf")
```

### Figure 2
# functional redundancy in taxonomic models
# working on code for this using Dan's output, but will need to rerun with new taxonomic models to make sure models are consistent (something probably changed between when Dan did it and when I last reran the models)
# Ceres currently under major maintenance and inaccessible, so need to wait to rerun taxonomic models
# also need the file that says what enzymes are in each taxon
```{r}
### hops 2018
hops2018 <- read.delim("picrust2_files/Hops.2018/EC/pred_metagenome_contrib.tsv",
                       sep = "\t")

# change name of function. column
colnames(hops2018)[2] <- "Hops2018_EC"

# filter down to function and taxon columns to reduce size of df
hops2018 <- hops2018[,2:3]

# group by taxon and make a list of ECs per taxon
hops2018 <- hops2018 %>% 
  group_by(taxon) %>% 
  summarise(Hops2018_ECs = list(Hops2018_EC))

### hops ARS
hopsARS <- read.delim("picrust2_files/Hops.ARS/EC/pred_metagenome_contrib.tsv.gz",
                       sep = "\t")
colnames(hopsARS)[2] <- "HopsARS_EC"
hopsARS <- hopsARS[,2:3]
hopsARS <- hopsARS %>% 
  group_by(taxon) %>% 
  summarise(HopsARS_ECs = list(HopsARS_EC))

### NRCS part 1
NRCS1 <- readRDS("picrust2_files/NRCS/EC/pred_metagenome_contrib1.RDS")
colnames(NRCS1)[2] <- "NRCS1_EC"
NRCS1 <- NRCS1[,2:3]
NRCS1 <- NRCS1 %>% 
  group_by(taxon) %>% 
  summarise(NRCS1_ECs = list(NRCS1_EC))

### NRCS part 2
NRCS2 <- readRDS("picrust2_files/NRCS/EC/pred_metagenome_contrib2.RDS")
colnames(NRCS2)[2] <- "NRCS2_EC"
NRCS2 <- NRCS2[,2:3]
NRCS2 <- NRCS2 %>% 
  group_by(taxon) %>% 
  summarise(NRCS2_ECs = list(NRCS2_EC))

### Rangeland
rangeland <- read.delim("picrust2_files/Rangeland/EC/pred_metagenome_contrib.tsv.gz",
                       sep = "\t")
colnames(rangeland)[2] <- "Rangeland_EC"
rangeland <- rangeland[,2:3]
rangeland <- rangeland %>% 
  group_by(taxon) %>% 
  summarise(Rangeland_ECs = list(Rangeland_EC))

### merge files and save
all_merged <- hops2018 %>% 
  full_join(hopsARS, by = "taxon") %>% 
  full_join(NRCS1, by = "taxon") %>% 
  full_join(NRCS2, by = "taxon") %>% 
  full_join(rangeland, by = "taxon")

saveRDS(all_merged, "picrust2_files/ECs_per_otu.RDS")
all_merged <- readRDS("picrust2_files/ECs_per_otu.RDS")

# merge lists together
all_merged <- all_merged %>% 
  mutate(all_ECs = mapply(c, Hops2018_ECs, HopsARS_ECs, NRCS1_ECs,
                          NRCS2_ECs, Rangeland_ECs))
all_merged_2 <- all_merged[,c(1,7)]


####### these OTU ids are output by picrust2. Figure out where those are coming from and try to connect them with taxa before continuing

# import varimps (will put into a loop later, just getting hard part of code down)
# importing Dan's old varimps, but will rerun when scinet is back up
ace_varimps <- read.csv("dan_out_2/ace_TAX_RA_varimp.csv")




######## 
# decide if I want to stick with this file, or figure out how to merge columns and OTU lists
# may have to just search for ECs in each column?

# this works for searching a list (if I need it)
list.search(hops_merged$ECs.x[1], . =="EC:1.1.1.1")

# try to merge lists for each OTU
#all_merged$ECs_all <- Map(function(...) sort(union(...)), all_merged$Hops2018_ECs, all_)

DF$result <- Map(function(...) sort(union(...)), DF$pos[1], DF$pos)


# figure out what file I can use to connect OTU IDs with taxa

```

### Figure 3
# PCOAs using tsne
# add PCOA of taxonomic data
# calculate permanova between land use and climate zone for EC and taxonomic data
```{r}
P <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')

gibbs <- read.csv('picrust2_files/GIBBs.EC.numbers.csv')
gibbs <- gibbs %>%
  filter(Missing != 'x')

data <- P %>%
  select(gibbs$EC)

meta <- P
meta <- meta[,c(1:2,2439:2587)]

p1 <- M3C::tsne(t(data), labels=FALSE, dotsize = 0, seed=1234, perplex=15) + theme_bw() +
  geom_point(aes(fill=meta$ClimateZ, shape=meta$Current_Land_Use), size=4) +
  scale_shape_manual(values=c(21,22,23,24), labels = c("Annual Cropland","Perennial Cropland","Rangeland")) +
  guides(size="none",
         fill = guide_legend(override.aes = list(size = 4, shape=21))) +
  labs(fill="Climate Zone", shape="Land Use", x="tSNE 1", y="tSNE 2")
p1
#ggsave(paste0("figures/All_GIBBs_RA_corr_tSNE.png"), device='png', dpi=600, height=4, width=6)

# permanova between land use and climate zone
permanova_land <- pairwise.adonis2(data ~ Current_Land_Use,
                                data = meta,
                                permutations = 9999)
# p-vals for all comparisons less than 0.05

permanova_climate <- pairwise.adonis2(data ~ ClimateZ,
                                data = meta,
                                permutations = 9999)
```

### Figure 4 and Figure 5
# Feature importance distributions of indicators with better all enzyme models
# separating by positively (Figure 4) and negatively (Figure 5) correlated enzymes
```{r}
myVars <- c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'zn', 'ace', 'activeC', 'ph', 'p', 'k','fe')

for (myVar in myVars) {
     filenames <- list.files(paste0("machine_learning/16S_EC/",myVar,"_model_results_RA_corr/"), pattern = paste0(myVar,"_RA_corr_varimp*"), full.names = TRUE)
  varimps <- lapply(filenames, read.csv, header = FALSE)
  varimps_full <- varimps %>% 
    reduce(full_join)
  
  names(varimps_full) <- c('Predictor', 'VIP', 'Run')
  varimps_full$Predictor <- gsub("EC_","EC:", varimps_full$Predictor)
  varimps_full$Predictor <- gsub("_","\\.", varimps_full$Predictor)

 gDF <- varimps_full %>%
    left_join(gibbs, by=c("Predictor"="EC")) %>% 
    mutate(Predictor=fct_reorder(Predictor, VIP, .fun=median, .desc=TRUE))

 # calculate positive and negative correlations between enzymes and indicators
 ml_EC_RA_corr <- readRDS('machine_learning/EC_RA_corr/ml_EC_RA_corr.RDS')
 
 # putting some indicators into their own loops since climateZ, clay, DNA not all found in them
 if (myVar == 'p') {
  ml_indic <- ml_EC_RA_corr %>% 
    select(gDF$Predictor, paste0(myVar)) %>% 
    subset(select=-c(DNA))
 }
 
 if (myVar == 'ph') {
  ml_indic <- ml_EC_RA_corr %>% 
    select(gDF$Predictor, paste0(myVar)) %>% 
    subset(select=-c(clay))
 }
 
 if (myVar == 'zn') {
  ml_indic <- ml_EC_RA_corr %>% 
    select(gDF$Predictor, paste0(myVar)) %>% 
    subset(select=-c(clay, DNA))
 }
 
 if (myVar == 'fe') {
  ml_indic <- ml_EC_RA_corr %>% 
    select(gDF$Predictor, paste0(myVar)) %>% 
    subset(select=-c(ClimateZ))
 }
 
 if (myVar %in% c('SOM', 'resp', 'agg_stab', 'water_cap', 'mg', 'mn', 'ace', 'activeC', 'k')) {
  ml_indic <- ml_EC_RA_corr %>% 
    select(gDF$Predictor, paste0(myVar)) %>% 
    subset(select=-c(ClimateZ, clay, DNA))
 }
 
 ml_indic <- as.matrix(ml_indic)
 ec.mat <- rcorr(ml_indic)
 ec.r <- ec.mat$r
 ec.indic <- ec.r %>% 
   subset(select = paste0(myVar))
 ec.indic <- data.frame(ec.indic)

 # get positive and negative correlation values
 ec.indic.pos <- ec.indic %>% 
   filter(ec.indic[,1] > 0)
 ec.indic.neg <- ec.indic %>% 
   filter(ec.indic[,1] < 0)
 ec.indic.pos <- rownames_to_column(ec.indic.pos, var = "Predictor")
 ec.indic.neg <- rownames_to_column(ec.indic.neg, var = "Predictor")
 ec.indic.pos <- ec.indic.pos[ec.indic.pos$Predictor != myVar,]
 ec.indic.neg <- ec.indic.neg[ec.indic.neg$Predictor != myVar,]

  # filter out poor indicators in pos and neg correlations
  # add this if I want to include climate, clay, and DNA:
  # | Predictor %in% c('clay','ClimateZ','DNA')
 gDF.pos <- gDF %>% 
   filter(Predictor %in% ec.indic.pos$Predictor) 
 gDF.neg <- gDF %>% 
   filter(Predictor %in% ec.indic.neg$Predictor) 
 
  temp.pos <- gDF.pos %>% group_by(Predictor) %>%
    summarise(Var = median(VIP)) %>%
    top_n(10, Var) 
  temp.neg <- gDF.neg %>% group_by(Predictor) %>%
    summarise(Var = median(VIP)) %>%
    top_n(10, Var) 
  
  gDF.pos <- gDF.pos %>%
    filter(Predictor %in% temp.pos$Predictor)
  gDF.neg <- gDF.neg %>%
    filter(Predictor %in% temp.neg$Predictor)
  
  colors <- c("Biocontrol" = "#1F77B4FF",
              "C cycling" = "#FF7F0EFF",
              "C/N cycling" = "#2CA02CFF",
              "N cycling" = "#D62728FF",
              "P cycling" = "#9467BDFF",
              "S cycling" = "#8C564BFF",
              "Stress" = "#E377C2FF",
              "Siderophore" = "#7F7F7FFF",
              "NA" = "white")
  
  # changing myVar format for cleaner plot titles
  if (myVar == 'resp'){
    myVar <- "Resp"
  }
  if (myVar == 'agg_stab'){
    myVar <- "AggStab"
  }
  if (myVar == 'water_cap'){
    myVar <- "WaterCap"
  }
  if (myVar == 'mg'){
    myVar <- "Mg"
  }
  if (myVar == 'mn'){
    myVar <- "Mn"
  }
  if (myVar == 'zn'){
    myVar <- "Zn"
  }
  if (myVar == 'ace'){
    myVar <- "ACE"
  }
  if (myVar == 'activeC'){
    myVar <- "ActiveC"
  }
  if (myVar == 'ph'){
    myVar <- "pH"
  }
  if (myVar == 'p'){
    myVar <- "P"
  }
  if (myVar == 'k'){
    myVar <- "K"
  }
  if (myVar == 'fe'){
    myVar <- "Fe"
  }
  # positive correlation graph
  means <- aggregate(VIP ~ Predictor, gDF.pos, median)
  means$Category <- 'Other'
  means$VIP <- round(means$VIP, 3)
  p.pos <- ggplot(gDF.pos, aes(x=Predictor, y=VIP, fill=Category)) +
    scale_fill_manual(values=colors, na.value="white") +
    geom_boxplot(alpha = 1, fatten = 1) + 
    theme_bw() +
    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5), 
          legend.position = "none", plot.title = element_text(size = 12)) +
    labs(x="", y="", title = myVar)
  p.pos
  assign(paste0(myVar,"_pos"),p.pos)
    ggsave(paste0("figures/varimps/", myVar, "_RA_corr_varimp.pos.png"), device='png', dpi=600, height=8, width=8)
  
  # negative correlation graph
  means <- aggregate(VIP ~ Predictor, gDF.neg, median)
  means$Category <- 'Other'
  means$VIP <- round(means$VIP, 3)
  p.neg <- ggplot(gDF.neg, aes(x=Predictor, y=VIP, fill=Category)) +
    scale_fill_manual(values=colors, na.value="white") +
    geom_boxplot(alpha = 1, fatten = 1) + 
    theme_bw() +
    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5), 
          legend.position = "none", plot.title = element_text(size = 12)) +
    labs(x="", y="", title = myVar)
  p.neg
  assign(paste0(myVar,"_neg"),p.neg)
    ggsave(paste0("figures/varimps/", myVar, "_RA_corr_varimp.neg.png"), device='png', dpi=600, height=8, width=8)
}

# create legend of all GIBBs colors to add to merged plot
p_legend <- ggplot(gibbs, aes(x = EC, fill = Category)) +
  scale_fill_manual(values = colors) +
  geom_boxplot()
p_legend
p_legend2 <- get_legend(p_legend)
p_legend3 <- as_ggplot(p_legend2)
p_legend3

# positive correlations graph
p_merged_pos <- ggarrange(ACE_pos, ActiveC_pos, AggStab_pos, 
                                Fe_pos, K_pos, Mg_pos, Mn_pos, 
                                P_pos, pH_pos, Resp_pos, 
                                SOM_pos, WaterCap_pos, Zn_pos,
                                p_legend3, nrow = 2, ncol = 7)
p_merged_pos

annotate_figure(p_merged_pos, left = text_grob("Variable Importance", rot = 90, vjust = 1, size = 14),
                bottom = text_grob("Positively Correlated Model Predictors", size = 14),
                top = text_grob("All Enzymes", size = 14))

ggsave("figures/all_enzyme_varimps_pos.png", device='png', dpi=600, height=6, width=12)

# negative correlations graph
p_merged_neg <- ggarrange(ACE_neg, ActiveC_neg, AggStab_neg, 
                                Fe_neg, K_neg, Mg_neg, Mn_neg, 
                                P_neg, pH_neg, Resp_neg, 
                                SOM_neg, WaterCap_neg, Zn_neg,
                                p_legend3, nrow = 2, ncol = 7)
p_merged_neg

annotate_figure(p_merged_neg, left = text_grob("Variable Importance", rot = 90, vjust = 1, size = 14),
                bottom = text_grob("Negatively Correlated Model Predictors", size = 14),
                top = text_grob("All Enzymes", size = 14))

ggsave("figures/all_enzyme_varimps_neg.png", device='png', dpi=600, height=6, width=12)
```







